// @iruka-edu/mini-game-sdk/dist/index.mjs
var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/index.ts
import {
  CONTRACT_VERSION
} from "@iruka-edu/game-core";
import {
  makeEnvelope as makeEnvelope2,
  isRecord as isRecord2,
  isValidEnvelope as isValidEnvelope2,
  isHubCommandType as isHubCommandType2,
  isGameEventType,
  clamp01 as clamp015,
  MAX_MESSAGE_BYTES as MAX_MESSAGE_BYTES2,
  estimateJsonBytes as estimateJsonBytes2,
  validateMessageSize as validateMessageSize2
} from "@iruka-edu/game-core";
import {
  connectToHub as connectToHub2,
  HandshakeHubBridge
} from "@iruka-edu/game-core";
import {
  validateManifest
} from "@iruka-edu/game-core";
import {
  normalizeResult
} from "@iruka-edu/game-core";

// src/core/game/baseGame.ts
import {
  connectToHub,
  ErrorCode
} from "@iruka-edu/game-core";

// src/runtime/ui/OverlayManager.ts
var OverlayManager = class {
  constructor() {
    __publicField(this, "container");
    this.container = document.createElement("div");
    this.container.id = "iruka-overlay-layer";
    Object.assign(this.container.style, {
      position: "absolute",
      top: "0",
      left: "0",
      width: "100%",
      height: "100%",
      pointerEvents: "none",
      // Quan trọng: Để click xuyên qua game lúc bình thường
      zIndex: "9999",
      display: "flex",
      justifyContent: "center",
      alignItems: "center"
    });
    document.body.appendChild(this.container);
  }
  showLoading() {
    this.container.innerHTML = `<div class="spinner" style="color: white; font-family: sans-serif;">Loading...</div>`;
    this.container.style.pointerEvents = "all";
    this.container.style.backgroundColor = "rgba(0,0,0,0.8)";
  }
  hide() {
    this.container.innerHTML = "";
    this.container.style.pointerEvents = "none";
    this.container.style.backgroundColor = "transparent";
  }
  showResult(score, stars, onRetry) {
    this.container.style.pointerEvents = "all";
    this.container.style.backgroundColor = "rgba(0,0,0,0.7)";
    this.container.innerHTML = `
      <div class="result-card" style="background: white; padding: 2rem; border-radius: 1rem; text-align: center; font-family: sans-serif; min-width: 300px;">
        <h2 style="margin: 0 0 1rem 0; color: #333;">Game Complete!</h2>
        <div class="stars" style="font-size: 2rem; margin-bottom: 1rem;">${"\u2B50".repeat(
      stars
    )}</div>
        <h1 style="font-size: 3rem; margin: 0 0 1.5rem 0; color: #4CAF50;">${score} pts</h1>
        <button id="btn-retry" style="padding: 10px 20px; font-size: 1.2rem; cursor: pointer; background: #2196F3; color: white; border: none; border-radius: 4px;">Ch\u01A1i l\u1EA1i</button>
      </div>
    `;
    document.getElementById("btn-retry")?.addEventListener("click", () => {
      this.hide();
      onRetry();
    });
  }
  showError(message, code) {
    this.container.style.pointerEvents = "all";
    this.container.style.backgroundColor = "rgba(50,0,0,0.9)";
    this.container.innerHTML = `
      <div style="color: white; padding: 20px; text-align: center;">
        <h1 style="color: #ff5252;">Game Error</h1>
        <p style="font-size: 1.2rem;">${message || "Unknown Error"}</p>
        <small style="color: #ccc;">${code || ""}</small>
      </div>
    `;
  }
};

// src/phaser/IrukaPlugin.ts
import Phaser from "phaser";
var IrukaPlugin = class extends Phaser.Plugins.BasePlugin {
  constructor(pluginManager) {
    super(pluginManager);
    __publicField(this, "bridge");
  }
  /**
   * Called by BaseGame after the game instance is created.
   * Injects the authenticated bridge into the plugin.
   */
  init(bridge) {
    this.bridge = bridge;
    console.log("[IrukaSDK] \u{1F50C} Phaser Plugin Attached");
  }
  // ==========================================
  // PUBLIC API (For Game Developers)
  // ==========================================
  /**
   * Get the current game context (User profile, Difficulty, Settings, etc.)
   */
  get context() {
    if (!this.bridge) {
      console.warn("[IrukaSDK] Accessing context before bridge is ready");
      return {};
    }
    return this.bridge.context;
  }
  /**
   * Submit the final game result.
   * @param score - The score achieved (number).
   * @param stars - (Optional) Star rating (0-3).
   * @param extraData - (Optional) Any custom data for analytics.
   */
  submitResult(score, stars = 0, extraData = {}) {
    this.bridge.send("EVT:COMPLETE", {
      score,
      completion: 1,
      // 100% completed
      details: {
        stars,
        ...extraData
      },
      timestamp: Date.now()
    });
  }
  /**
   * Save the current game state (e.g., current level, inventory).
   * @param data - The object to save.
   */
  saveState(data) {
    this.bridge.send("CMD:SAVE_STATE", data);
  }
  /**
   * Request the Hub to play a haptic feedback (if supported).
   */
  vibrate(duration = 200) {
    if (navigator.vibrate) {
      navigator.vibrate(duration);
    }
  }
  // ==========================================
  // LEGACY API (Backward Compatibility)
  // ==========================================
  /**
   * Submit score to hub
   * @deprecated Use submitResult() instead
   */
  submitScore(score, delta) {
    console.warn("[IrukaSDK] submitScore() is deprecated, use submitResult() instead");
    this.bridge.send("EVT:SCORE_UPDATE", {
      score,
      delta,
      timestamp: Date.now()
    });
  }
  /**
   * Submit progress update
   * @deprecated Use bridge directly or submitResult() for final completion
   */
  submitProgress(current, total) {
    console.warn("[IrukaSDK] submitProgress() is deprecated");
    this.bridge.send("EVT:PROGRESS", {
      current,
      total,
      percentage: current / total,
      timestamp: Date.now()
    });
  }
  /**
   * Complete game with final result
   * @deprecated Use submitResult() instead
   */
  completeGame(result) {
    console.warn("[IrukaSDK] completeGame() is deprecated, use submitResult() instead");
    this.submitResult(result.score, 0, { timeMs: result.timeMs, ...result.extras });
  }
  /**
   * Load saved game state
   * @deprecated Use saveState() instead
   */
  async loadState() {
    console.warn("[IrukaSDK] loadState() is deprecated");
    this.bridge.send("CMD:REQUEST_LOAD");
    return null;
  }
  /**
   * Get launch context (readonly)
   * @deprecated Use context property instead
   */
  getContext() {
    console.warn("[IrukaSDK] getContext() is deprecated, use context property instead");
    return this.context;
  }
  /**
   * Send telemetry event
   * @deprecated Use bridge directly
   */
  track(eventType, data) {
    console.warn("[IrukaSDK] track() is deprecated");
    this.bridge.send("EVT:TELEMETRY", {
      eventType,
      data,
      timestamp: Date.now()
    });
  }
};
var IRUKA_PLUGIN_KEY = "IrukaPlugin";
function createIrukaPluginConfig() {
  return {
    key: IRUKA_PLUGIN_KEY,
    plugin: IrukaPlugin,
    start: true,
    mapping: "iruka"
  };
}

// src/core/game/baseGame.ts
var BaseGame = class {
  constructor() {
    __publicField(this, "bridge");
    __publicField(this, "context");
    __publicField(this, "ui");
    this.ui = new OverlayManager();
    this.ui.showLoading();
    this.bootstrap().catch((err2) => {
      console.error("\u2620\uFE0F Game Bootstrap Failed:", err2);
      this.ui.showError(err2.message, err2.code);
    });
  }
  async bootstrap() {
    console.log("\u{1F680} Starting Game Bootstrap...");
    try {
      this.bridge = await connectToHub();
      this.context = this.bridge.context;
      console.log("\u2705 Connected to Hub. Mode:", this.context.mode);
      await this.onInit(this.context);
      this.initializePhaserPlugin();
      this.ui.hide();
      await this.onStart();
    } catch (error) {
      if (error?.code === ErrorCode.SYS_INVALID_CONTEXT) {
        console.error("Hub sent bad data!");
      }
      throw error;
    }
  }
  /**
   * Initialize Phaser plugin if a Phaser game instance is available
   * This allows scenes to access SDK via this.iruka
   */
  initializePhaserPlugin() {
    if (typeof window !== "undefined" && window.phaserGame) {
      const game = window.phaserGame;
      if (game.isBooted) {
        this.injectPluginToPhaserGame(game);
      } else {
        game.events.once("ready", () => {
          this.injectPluginToPhaserGame(game);
        });
      }
    }
  }
  /**
   * Inject the bridge into the Phaser plugin
   */
  injectPluginToPhaserGame(game) {
    const plugin = game.plugins.get(IRUKA_PLUGIN_KEY);
    if (plugin) {
      plugin.init(this.bridge);
      console.log("\u2705 Phaser IrukaPlugin initialized with bridge");
    } else {
      console.warn("\u26A0\uFE0F IrukaPlugin not found in Phaser game. Make sure to include it in your game config.");
    }
  }
  /**
   * Helper to merge default SDK plugins with Developer's config.
   * Game Devs should call this if they override getGameConfig.
   */
  getStandardPhaserConfig(overrides) {
    const plugins = overrides.plugins || {};
    const globalPlugins = Array.isArray(plugins) ? plugins : plugins.global || [];
    return {
      ...overrides,
      plugins: {
        ...plugins,
        global: [
          ...globalPlugins,
          {
            key: IRUKA_PLUGIN_KEY,
            plugin: IrukaPlugin,
            start: true,
            mapping: "iruka"
            // <--- MAGIC: Defines "this.iruka" in Scenes
          }
        ]
      }
    };
  }
};

// src/react/IrukaGameHost.tsx
import { useEffect, useRef } from "react";
import { jsx } from "react/jsx-runtime";
function IrukaGameHost({ createGame, config, className }) {
  const gameRef = useRef(null);
  useEffect(() => {
    const game = createGame();
    gameRef.current = game;
    (async () => {
      await game.onInit(config);
      await game.onStart();
    })();
    const onVis = () => document.hidden ? game.onPause?.() : game.onResume?.();
    document.addEventListener("visibilitychange", onVis);
    return () => {
      document.removeEventListener("visibilitychange", onVis);
      game.onDispose?.();
      gameRef.current = null;
    };
  }, [config, createGame]);
  return /* @__PURE__ */ jsx("div", { id: "game-root", className: className ?? "w-full h-full" });
}

// src/core/game/index.ts
var game_exports = {};
__export(game_exports, {
  addHint: () => addHint,
  createGameSdk: () => createGameSdk,
  finalizeAttempt: () => finalizeAttempt,
  finishQuestionTimer: () => finishQuestionTimer,
  getStatsSnapshot: () => getStatsSnapshot,
  prepareSubmitData: () => prepareSubmitData,
  recordCorrect: () => recordCorrect,
  recordQuestionDuration: () => recordQuestionDuration,
  recordWrong: () => recordWrong,
  resetAll: () => resetAll,
  retryFromStart: () => retryFromStart,
  setPassCriteria: () => setPassCriteria,
  setTotal: () => setTotal,
  startQuestionTimer: () => startQuestionTimer
});

// src/core/protocol/protocol.ts
function makeEnvelope(params) {
  return {
    sdkVersion: params.sdkVersion,
    source: params.source,
    sessionId: params.sessionId,
    gameId: params.gameId,
    lessonId: params.lessonId,
    gameVersion: params.gameVersion,
    type: params.type,
    payload: params.payload
  };
}
function estimateJsonBytes(value) {
  try {
    return JSON.stringify(value).length;
  } catch {
    return Number.POSITIVE_INFINITY;
  }
}
function isRecord(v) {
  return typeof v === "object" && v !== null;
}
function isValidEnvelope(v) {
  if (!isRecord(v)) return false;
  if (typeof v.sdkVersion !== "string" || v.sdkVersion.length < 1) return false;
  if (v.source !== "hub" && v.source !== "game") return false;
  if (typeof v.sessionId !== "string" || v.sessionId.length < 1) return false;
  if (typeof v.gameId !== "string" || v.gameId.length < 1) return false;
  if (typeof v.lessonId !== "string" || v.lessonId.length < 1) return false;
  if (typeof v.type !== "string" || v.type.length < 1) return false;
  return true;
}
function isHubCommandType(t) {
  return t === "INIT" || t === "START" || t === "PAUSE" || t === "RESUME" || t === "RESIZE" || t === "QUIT" || t === "SET_STATE";
}
function clamp01(n) {
  if (!Number.isFinite(n)) return 0;
  if (n < 0) return 0;
  if (n > 1) return 1;
  return n;
}
function err(msg) {
  return { ok: false, error: msg };
}
function ok(value) {
  return { ok: true, value };
}
function isNumber(v) {
  return typeof v === "number" && Number.isFinite(v);
}
function isString(v) {
  return typeof v === "string";
}
function isBoolean(v) {
  return typeof v === "boolean";
}
function getPath(obj, key) {
  return obj[key];
}
var MAX_MESSAGE_BYTES = 16 * 1024;
function validateMessageSize(envelope) {
  const bytes = estimateJsonBytes(envelope);
  if (bytes > MAX_MESSAGE_BYTES) {
    return err(`payload_too_large: ${bytes} bytes > ${MAX_MESSAGE_BYTES}`);
  }
  return ok(void 0);
}
function validateLaunchContext(v) {
  if (!isRecord(v)) return err("LaunchContext must be an object");
  const sessionId = getPath(v, "sessionId");
  if (!isString(sessionId) || sessionId.length < 1)
    return err("LaunchContext.sessionId must be a non-empty string");
  const lessonId = getPath(v, "lessonId");
  if (!isString(lessonId) || lessonId.length < 1)
    return err("LaunchContext.lessonId must be a non-empty string");
  const gameId = getPath(v, "gameId");
  if (!isString(gameId) || gameId.length < 1)
    return err("LaunchContext.gameId must be a non-empty string");
  const gameVersion = getPath(v, "gameVersion");
  if (!isString(gameVersion) || gameVersion.length < 1)
    return err("LaunchContext.gameVersion must be a non-empty string");
  const locale = getPath(v, "locale");
  if (!isString(locale) || locale.length < 1)
    return err("LaunchContext.locale must be a non-empty string");
  const sdkVersion = getPath(v, "sdkVersion");
  if (sdkVersion !== void 0 && !isString(sdkVersion))
    return err("LaunchContext.sdkVersion must be a string");
  const playerId = getPath(v, "playerId");
  if (playerId !== void 0 && !isString(playerId))
    return err("LaunchContext.playerId must be a string");
  const launchToken = getPath(v, "launchToken");
  if (launchToken !== void 0 && !isString(launchToken))
    return err("LaunchContext.launchToken must be a string");
  const difficulty = getPath(v, "difficulty");
  if (difficulty !== void 0 && !isString(difficulty))
    return err("LaunchContext.difficulty must be a string");
  const seed = getPath(v, "seed");
  if (seed !== void 0 && !isNumber(seed))
    return err("LaunchContext.seed must be a number");
  const profile = getPath(v, "profile");
  if (profile !== void 0 && !isRecord(profile))
    return err("LaunchContext.profile must be an object");
  const sizeHint = getPath(v, "sizeHint");
  if (sizeHint !== void 0) {
    if (!isRecord(sizeHint))
      return err("LaunchContext.sizeHint must be an object");
    const width = getPath(sizeHint, "width");
    const height = getPath(sizeHint, "height");
    const dpr = getPath(sizeHint, "devicePixelRatio");
    if (!isNumber(width) || width <= 0)
      return err("LaunchContext.sizeHint.width must be a positive number");
    if (!isNumber(height) || height <= 0)
      return err("LaunchContext.sizeHint.height must be a positive number");
    if (dpr !== void 0 && (!isNumber(dpr) || dpr <= 0 || dpr > 8)) {
      return err("LaunchContext.sizeHint.devicePixelRatio must be (0..8]");
    }
  }
  const featureFlags = getPath(v, "featureFlags");
  if (featureFlags !== void 0) {
    if (!isRecord(featureFlags))
      return err("LaunchContext.featureFlags must be an object");
    for (const [k, val] of Object.entries(featureFlags)) {
      if (!isBoolean(val))
        return err(`LaunchContext.featureFlags['${k}'] must be boolean`);
    }
  }
  return ok({
    sdkVersion,
    playerId,
    sessionId,
    lessonId,
    gameId,
    gameVersion,
    locale,
    difficulty,
    seed,
    launchToken,
    profile,
    sizeHint,
    featureFlags
  });
}
function validateResizePayload(v) {
  if (!isRecord(v)) return err("RESIZE.payload must be an object");
  const width = getPath(v, "width");
  const height = getPath(v, "height");
  const dpr = getPath(v, "dpr");
  if (!isNumber(width) || width <= 0)
    return err("RESIZE.payload.width must be a positive number");
  if (!isNumber(height) || height <= 0)
    return err("RESIZE.payload.height must be a positive number");
  if (!isNumber(dpr) || dpr <= 0 || dpr > 8)
    return err("RESIZE.payload.dpr must be (0..8]");
  return ok({ width, height, dpr });
}
function validateIncomingEnvelope(v) {
  if (!isValidEnvelope(v)) return err("Invalid envelope");
  const sizeOk = validateMessageSize(v);
  if (!sizeOk.ok) return sizeOk;
  return ok(v);
}
function parseHubCommandFromEnvelope(env) {
  const t = env.type;
  if (!isHubCommandType(t)) return err(`Unknown HubCommand type: ${t}`);
  switch (t) {
    case "INIT": {
      const res = validateLaunchContext(env.payload);
      if (!res.ok) return err(`INIT.payload invalid: ${res.error}`);
      return ok({ type: "INIT", payload: res.value });
    }
    case "RESIZE": {
      const res = validateResizePayload(env.payload);
      if (!res.ok) return err(`RESIZE.payload invalid: ${res.error}`);
      return ok({ type: "RESIZE", payload: res.value });
    }
    case "START":
    case "PAUSE":
    case "RESUME":
    case "QUIT":
      return ok({ type: t });
    case "SET_STATE": {
      return ok({ type: "SET_STATE", payload: env.payload });
    }
  }
}

// src/core/shared/sdkContext.ts
var _fallback = {};
var _override = {};
function genLocalSessionId() {
  const c = typeof globalThis !== "undefined" ? globalThis.crypto : void 0;
  if (c?.randomUUID) return `local-${c.randomUUID()}`;
  const t = Date.now().toString(36);
  const r = Math.random().toString(36).slice(2, 10);
  return `local-${t}-${r}`;
}
function configureSdkContext(opts) {
  if (!opts.fallback) return;
  _fallback = { ..._fallback, ...opts.fallback };
  if (!_fallback.sessionId) {
    _fallback.sessionId = genLocalSessionId();
  }
}
function setSdkContext(partial) {
  _override = { ..._override, ...partial };
}
function getSdkContext() {
  return { ..._fallback, ..._override };
}
function resetSdkContext() {
  _fallback = {};
  _override = {};
}

// src/core/shared/sdkSpy.ts
var cfg = {
  maxRecords: 2e3,
  maxJsonChars: 4e3
};
var enabled = false;
var records = [];
function safeCloneSmall(v, maxChars = cfg.maxJsonChars) {
  try {
    const json = JSON.stringify(v, (_k, val) => {
      if (val instanceof Error) {
        return { name: val.name, message: val.message, stack: val.stack };
      }
      return val;
    });
    if (json.length <= maxChars) return JSON.parse(json);
    return { truncated: true, preview: json.slice(0, maxChars) + "\u2026" };
  } catch {
    return { unserializable: true, preview: String(v) };
  }
}
function spyPush(cat, name, data) {
  if (!enabled) return;
  records.unshift({
    t: performance.now(),
    cat,
    name,
    data: data === void 0 ? void 0 : safeCloneSmall(data)
  });
  if (records.length > cfg.maxRecords) records.length = cfg.maxRecords;
}
var __testSpy = {
  enable() {
    enabled = true;
  },
  disable() {
    enabled = false;
  },
  isEnabled() {
    return enabled;
  },
  reset() {
    records.length = 0;
  },
  getRecords() {
    return records.map((r) => ({ ...r, data: r.data }));
  },
  getSummary() {
    const out = {};
    for (const r of records) {
      const k = `${r.cat}:${r.name}`;
      out[k] = (out[k] || 0) + 1;
    }
    return out;
  },
  setConfig(patch) {
    if (typeof patch.maxRecords === "number" && patch.maxRecords > 0) cfg.maxRecords = Math.floor(patch.maxRecords);
    if (typeof patch.maxJsonChars === "number" && patch.maxJsonChars > 100) cfg.maxJsonChars = Math.floor(patch.maxJsonChars);
  }
};

// src/core/game/gameSdk.v1.ts
function normalizeOrigin(origin) {
  return origin.replace(/\/+$/, "");
}
var DEV = typeof window !== "undefined" && window.__DEV__ || typeof location !== "undefined" && (location.hostname === "localhost" || location.hostname === "127.0.0.1");
function inferHubOrigin(fallback) {
  try {
    if (typeof document !== "undefined" && document.referrer) {
      return new URL(document.referrer).origin;
    }
  } catch {
  }
  return fallback;
}
function safePostToParent(env, targetOrigin) {
  window.parent.postMessage(env, DEV ? "*" : targetOrigin);
}
function createGameSdk(opts) {
  const hubOrigin = normalizeOrigin(inferHubOrigin(opts.hubOrigin));
  const allow = DEV ? null : new Set((opts.allowlistHubOrigins ?? [hubOrigin]).map(normalizeOrigin));
  let destroyed = false;
  let sessionId = null;
  let sdkVersion;
  let gameId;
  let lessonId;
  let gameVersion;
  let loadingWindowStart = 0;
  let loadingCount = 0;
  let scoreFlushTimer = null;
  let pendingScore = null;
  function invalid(reason, origin, data) {
    opts.onInvalidMessage?.({ reason, origin, data });
  }
  function sendEvent(type, payload) {
    if (destroyed) return;
    if (!sessionId) {
      return;
    }
    const env = makeEnvelope({
      sdkVersion,
      source: "game",
      sessionId,
      gameId,
      lessonId,
      gameVersion,
      type,
      payload
    });
    const bytes = estimateJsonBytes(env);
    spyPush("sdk", "sendEvent", { type, bytes });
    if (bytes > MAX_MESSAGE_BYTES) {
      spyPush("sdk", "payloadTooLarge", { type, bytes, limit: MAX_MESSAGE_BYTES });
      const tiny = makeEnvelope({
        sdkVersion,
        source: "game",
        sessionId,
        gameId,
        lessonId,
        gameVersion,
        type: "ERROR",
        payload: {
          code: "RUNTIME_EXCEPTION",
          message: `payload_too_large: ${bytes} bytes`
        }
      });
      safePostToParent(tiny, hubOrigin);
      return;
    }
    safePostToParent(env, hubOrigin);
  }
  const ERROR_MAX_PER_SESSION = 10;
  const ERROR_MIN_INTERVAL_MS = 400;
  const ERROR_DEDUPE_WINDOW_MS = 5e3;
  const ERROR_BUFFER_MAX = 5;
  let errorSentCount = 0;
  let lastErrorSentAt = 0;
  const recentErrorKeys = /* @__PURE__ */ new Map();
  const bufferedErrors = [];
  function normalizeErrorMessage(msg) {
    const s = typeof msg === "string" ? msg : msg ? String(msg) : "Unknown error";
    return s.length > 500 ? s.slice(0, 500) + "\u2026" : s;
  }
  function safeSerializeDetail(detail, maxLen = 2e3) {
    try {
      const json = JSON.stringify(detail, (_k, v) => {
        if (v instanceof Error) {
          return { name: v.name, message: v.message, stack: v.stack };
        }
        return v;
      });
      if (json.length <= maxLen) return JSON.parse(json);
      return { truncated: true, preview: json.slice(0, maxLen) + "\u2026" };
    } catch {
      return { unserializable: true, preview: normalizeErrorMessage(detail) };
    }
  }
  function makeErrorKey(code, message, detail) {
    const d = detail ?? {};
    const loc = `${d?.filename ?? ""}:${d?.lineno ?? ""}:${d?.colno ?? ""}`;
    return `${String(code)}|${message}|${loc}`;
  }
  function canSendErrorNow(key) {
    const now = performance.now();
    if (errorSentCount >= ERROR_MAX_PER_SESSION) return false;
    if (now - lastErrorSentAt < ERROR_MIN_INTERVAL_MS) return false;
    const last = recentErrorKeys.get(key);
    if (last && now - last < ERROR_DEDUPE_WINDOW_MS) return false;
    return true;
  }
  function markErrorSent(key) {
    const now = performance.now();
    lastErrorSentAt = now;
    errorSentCount += 1;
    recentErrorKeys.set(key, now);
    if (recentErrorKeys.size > 50) {
      for (const [k, t] of recentErrorKeys) {
        if (now - t > ERROR_DEDUPE_WINDOW_MS) recentErrorKeys.delete(k);
      }
    }
  }
  function reportError(code, message, detail) {
    try {
      const msg = normalizeErrorMessage(message);
      const safeDetail = detail ? safeSerializeDetail(detail) : void 0;
      const key = makeErrorKey(code, msg, safeDetail);
      if (!sessionId) {
        if (bufferedErrors.length < ERROR_BUFFER_MAX) {
          bufferedErrors.push({ code, message: msg, detail: safeDetail });
        }
        return;
      }
      if (!canSendErrorNow(key)) return;
      markErrorSent(key);
      sendEvent("ERROR", {
        code,
        message: msg,
        detail: safeDetail
      });
    } catch {
    }
  }
  function flushBufferedErrorsAfterInit() {
    if (!sessionId) return;
    if (bufferedErrors.length === 0) return;
    for (const e of bufferedErrors.splice(0)) {
      reportError(e.code, e.message, e.detail);
    }
  }
  function onMessage(event) {
    if (destroyed) return;
    const envRes0 = validateIncomingEnvelope(event.data);
    if (!envRes0.ok) return;
    const origin = normalizeOrigin(event.origin);
    if (!DEV && allow && !allow.has(origin)) {
      invalid("origin_not_allowed", origin, event.data);
      return;
    }
    if (event.source !== window.parent) {
      invalid("source_not_parent", origin, event.data);
      return;
    }
    const env = envRes0.value;
    const cmdRes = parseHubCommandFromEnvelope(env);
    if (!cmdRes.ok) {
      invalid(`invalid_hub_command: ${cmdRes.error}`, origin, event.data);
      return;
    }
    const cmd = cmdRes.value;
    spyPush("hub", "recvCommand", { type: cmd.type });
    if (cmd.type === "INIT") {
      sdkVersion = env.sdkVersion;
      sessionId = env.sessionId;
      gameId = env.gameId;
      lessonId = env.lessonId;
      gameVersion = env.gameVersion;
      spyPush("hub", "init", { sessionId });
      setSdkContext({
        sdkVersion,
        sessionId,
        gameId,
        lessonId,
        gameVersion,
        hubOrigin
      });
      if (cmd.payload?.sessionId && cmd.payload.sessionId !== sessionId) {
        invalid("sessionId_mismatch_envelope_vs_payload", origin, event.data);
        return;
      }
      if (cmd.payload?.gameId && cmd.payload.gameId !== gameId) {
        invalid("gameId_mismatch_envelope_vs_payload", origin, event.data);
        return;
      }
      if (cmd.payload?.lessonId && cmd.payload.lessonId !== lessonId) {
        invalid("lessonId_mismatch_envelope_vs_payload", origin, event.data);
        return;
      }
      errorSentCount = 0;
      lastErrorSentAt = 0;
      recentErrorKeys.clear();
      opts.onInit?.(cmd.payload);
      flushBufferedErrorsAfterInit();
      return;
    }
    if (!sessionId) {
      invalid("command_before_init", origin, event.data);
      return;
    }
    if (env.sessionId !== sessionId) {
      invalid("sessionId_mismatch", origin, event.data);
      return;
    }
    switch (cmd.type) {
      case "START":
        opts.onStart?.();
        break;
      case "PAUSE":
        opts.onPause?.();
        break;
      case "RESUME":
        opts.onResume?.();
        break;
      case "RESIZE":
        opts.onResize?.(cmd.payload);
        break;
      case "QUIT":
        opts.onQuit?.();
        break;
      case "SET_STATE":
        opts.onSetState?.(cmd.payload);
        break;
    }
  }
  window.addEventListener("message", onMessage);
  const onUnhandledRejection = (e) => {
    reportError(
      "UNHANDLED_REJECTION",
      "Unhandled promise rejection",
      { reason: e.reason }
    );
  };
  const onWindowError = (e) => {
    const isResourceError = e?.target && !e?.error;
    if (isResourceError) {
      const target = e.target;
      reportError(
        "RESOURCE_LOAD_ERROR",
        "Resource load error",
        {
          tagName: target?.tagName,
          src: target?.src,
          href: target?.href
        }
      );
      return;
    }
    reportError(
      "RUNTIME_EXCEPTION",
      e.message || "Window error",
      {
        filename: e.filename,
        lineno: e.lineno,
        colno: e.colno,
        stack: e.error && e.error.stack ? String(e.error.stack) : void 0
      }
    );
  };
  window.addEventListener("unhandledrejection", onUnhandledRejection);
  window.addEventListener("error", onWindowError, true);
  const api = {
    ready(payload) {
      sendEvent("READY", payload);
    },
    loading(progress01) {
      const now = performance.now();
      if (now - loadingWindowStart >= 1e3) {
        loadingWindowStart = now;
        loadingCount = 0;
      }
      if (loadingCount >= 10) return;
      loadingCount += 1;
      sendEvent("LOADING", { progress: clamp01(progress01) });
    },
    score(score, delta) {
      pendingScore = { score, delta };
      if (scoreFlushTimer) return;
      scoreFlushTimer = window.setTimeout(() => {
        scoreFlushTimer = null;
        if (!pendingScore) return;
        const last = pendingScore;
        pendingScore = null;
        sendEvent("SCORE_UPDATE", last);
      }, 100);
    },
    progress(payload) {
      sendEvent("PROGRESS", payload);
    },
    complete(payload) {
      sendEvent("COMPLETE", payload);
    },
    error(payload) {
      sendEvent("ERROR", payload);
    },
    destroy() {
      if (destroyed) return;
      destroyed = true;
      if (scoreFlushTimer) window.clearTimeout(scoreFlushTimer);
      scoreFlushTimer = null;
      pendingScore = null;
      window.removeEventListener("message", onMessage);
      window.removeEventListener("unhandledrejection", onUnhandledRejection);
      window.removeEventListener("error", onWindowError);
    },
    getSessionId() {
      return sessionId;
    },
    requestSave(payload) {
      sendEvent("REQUEST_SAVE", payload);
    },
    requestLoad() {
      sendEvent("REQUEST_LOAD");
    }
  };
  return api;
}

// src/core/game/statsCore.ts
function clamp012(n) {
  if (!Number.isFinite(n)) return 0;
  return Math.max(0, Math.min(1, n));
}
var passCriteria = {
  accuracyThreshold: 0.6,
  // default (ví dụ)
  requireCompletion: true
};
var state = {
  hintCount: 0,
  retryCount: 0,
  sessionMistakeCount: 0,
  total: 0,
  finalScore: 0,
  bestScore: 0,
  attemptCorrect: 0,
  attemptWrong: 0,
  attempts: [],
  attemptStartedAtMs: performance.now(),
  questionDurationsMs: [],
  currentQuestionStartedAtMs: void 0
};
function getAttemptIndex() {
  return state.attempts.length + 1;
}
function computeAccuracy() {
  const c = state.attemptCorrect;
  const w = state.attemptWrong;
  const denom = c + w;
  return denom <= 0 ? 0 : c / denom;
}
function computeCompletion() {
  if (state.total <= 0) return 0;
  return clamp012(state.attemptCorrect / state.total);
}
function snapshotAttemptPartial() {
  return {
    attemptIndex: getAttemptIndex(),
    correct: state.attemptCorrect,
    wrong: state.attemptWrong,
    total: state.total,
    completion: computeCompletion(),
    accuracy: computeAccuracy(),
    finalScore: state.finalScore,
    startedAtMs: state.attemptStartedAtMs,
    outcome: state.attempts[state.attempts.length - 1]?.outcome,
    questionDurationsMs: [...state.questionDurationsMs]
  };
}
function nowMs() {
  return performance.now();
}
function startQuestionTimer() {
  if (typeof state.currentQuestionStartedAtMs === "number") {
    finishQuestionTimer();
  }
  state.currentQuestionStartedAtMs = nowMs();
  spyPush("stats", "startQuestionTimer", { ok: true });
}
function finishQuestionTimer() {
  const start = state.currentQuestionStartedAtMs;
  if (typeof start !== "number") {
    spyPush("stats", "finishQuestionTimer", { ok: false, reason: "not_started" });
    return null;
  }
  const d = Math.max(0, nowMs() - start);
  state.questionDurationsMs.push(d);
  state.currentQuestionStartedAtMs = void 0;
  spyPush("stats", "finishQuestionTimer", {
    ok: true,
    durationMs: d,
    count: state.questionDurationsMs.length
  });
  return d;
}
function recordQuestionDuration(durationMs) {
  const d = Number.isFinite(durationMs) ? Math.max(0, durationMs) : 0;
  state.questionDurationsMs.push(d);
}
function setTotal(total) {
  state.total = Math.max(0, Math.floor(total));
  spyPush("stats", "setTotal", { total: state.total });
}
function addHint() {
  state.hintCount += 1;
  spyPush("stats", "addHint", { hintCount: state.hintCount });
}
function recordWrong() {
  state.attemptWrong += 1;
  state.sessionMistakeCount += 1;
  spyPush("stats", "recordWrong", {
    attemptWrong: state.attemptWrong,
    sessionMistakeCount: state.sessionMistakeCount
  });
}
function recordCorrect(opts) {
  state.attemptCorrect += 1;
  if (typeof opts?.scoreAbs === "number" && Number.isFinite(opts.scoreAbs)) {
    state.finalScore = opts.scoreAbs;
  } else if (typeof opts?.scoreDelta === "number" && Number.isFinite(opts.scoreDelta)) {
    state.finalScore += opts.scoreDelta;
  }
  if (state.finalScore > state.bestScore) {
    state.bestScore = state.finalScore;
  }
  spyPush("stats", "recordCorrect", {
    attemptCorrect: state.attemptCorrect,
    finalScore: state.finalScore,
    bestScore: state.bestScore,
    opts
  });
}
function retryFromStart() {
  state.retryCount += 1;
  state.attemptCorrect = 0;
  state.attemptWrong = 0;
  state.finalScore = 0;
  state.attemptStartedAtMs = performance.now();
  state.questionDurationsMs = [];
  state.currentQuestionStartedAtMs = void 0;
  spyPush("stats", "retryFromStart", { retryCount: state.retryCount });
}
function setPassCriteria(c) {
  if (typeof c.accuracyThreshold === "number") {
    const v = Math.max(0, Math.min(1, c.accuracyThreshold));
    passCriteria.accuracyThreshold = v;
  }
  if (typeof c.requireCompletion === "boolean") {
    passCriteria.requireCompletion = c.requireCompletion;
  }
}
function decidePassFail() {
  const a = snapshotAttemptPartial();
  const doneOk = passCriteria.requireCompletion ? a.completion >= 1 : true;
  const accOk = a.accuracy >= passCriteria.accuracyThreshold;
  return doneOk && accOk ? "pass" : "fail";
}
function finalizeAttempt(outcome) {
  if (typeof state.currentQuestionStartedAtMs === "number") {
    finishQuestionTimer();
  }
  const a = snapshotAttemptPartial();
  a.endedAtMs = performance.now();
  if (outcome === "quit") {
    a.outcome = "quit";
  } else if (outcome === "pass" || outcome === "fail") {
    a.outcome = outcome;
  } else {
    a.outcome = decidePassFail();
  }
  state.attempts.push(a);
  spyPush("stats", "finalizeAttempt", { outcome: state.attempts[state.attempts.length - 1]?.outcome });
}
function getStatsSnapshot() {
  return {
    // session counters
    hintCount: state.hintCount,
    retryCount: state.retryCount,
    sessionMistakeCount: state.sessionMistakeCount,
    // attempt metrics hiện tại
    finalScore: state.finalScore,
    bestScore: state.bestScore,
    accuracy: computeAccuracy(),
    completion: computeCompletion(),
    // chi tiết mistakes attempt
    attemptMistakeCount: state.attemptWrong,
    // history json theo attempt
    attempts: [...state.attempts],
    // debug / optional
    currentAttempt: snapshotAttemptPartial()
  };
}
function resetAll() {
  state.hintCount = 0;
  state.retryCount = 0;
  state.sessionMistakeCount = 0;
  state.total = 0;
  state.finalScore = 0;
  state.bestScore = 0;
  state.attemptCorrect = 0;
  state.attemptWrong = 0;
  state.attempts = [];
  state.attemptStartedAtMs = performance.now();
  spyPush("stats", "resetAll");
}
function prepareSubmitData() {
  const snap = getStatsSnapshot();
  const payload = {
    finalScore: snap.finalScore,
    // Điểm số cuối cùng
    accuracy: snap.accuracy,
    // Độ chính xác hiện tại (lần cuối lần thử)
    bestScore: snap.bestScore,
    // Điểm cao nhất
    completion: snap.completion,
    // Tiến độ hoàn thành hiện tại
    completion_history: snap.attempts,
    // Lịch sử tiến độ theo từng lần thử
    hintCount: snap.hintCount,
    // Số lần sử dụng gợi ý
    mistakeCount: snap.sessionMistakeCount,
    // Số lần mắc lỗi
    retryCount: snap.retryCount,
    // Số lần thử lại
    outcome: snap.currentAttempt?.outcome
  };
  spyPush("stats", "prepareSubmitData", {
    // đừng push cả completion_history nếu sợ vượt size
    finalScore: payload.finalScore,
    accuracy: payload.accuracy,
    bestScore: payload.bestScore,
    completion: payload.completion,
    hintCount: payload.hintCount,
    mistakeCount: payload.mistakeCount,
    retryCount: payload.retryCount,
    outcome: payload.outcome,
    attemptsCount: snap.attempts.length
  });
  return payload;
}

// src/core/runtime/index.ts
var runtime_exports = {};
__export(runtime_exports, {
  ResultOverlayDom: () => ResultOverlayDom,
  ScoreTimer: () => ScoreTimer,
  ScoreTimerHudDom: () => ScoreTimerHudDom,
  StateClient: () => StateClient,
  createGameSessionUiRuntime: () => createGameSessionUiRuntime
});

// src/core/runtime/score/scoreTimer.ts
var ScoreTimer = class {
  constructor() {
    __publicField(this, "state", {
      running: false,
      startAtMs: null,
      elapsedMs: 0,
      score: 0
    });
  }
  start(nowMs2 = performance.now()) {
    if (this.state.running) return;
    this.state.running = true;
    this.state.startAtMs = nowMs2;
  }
  pause(nowMs2 = performance.now()) {
    if (!this.state.running) return;
    this.state.elapsedMs = this.getElapsedMs(nowMs2);
    this.state.running = false;
    this.state.startAtMs = null;
  }
  reset() {
    this.state.running = false;
    this.state.startAtMs = null;
    this.state.elapsedMs = 0;
    this.state.score = 0;
  }
  setScore(score) {
    this.state.score = Math.max(0, Math.floor(score));
  }
  addScore(delta) {
    this.setScore(this.state.score + delta);
  }
  getScore() {
    return this.state.score;
  }
  getElapsedMs(nowMs2 = performance.now()) {
    if (!this.state.running || this.state.startAtMs == null) return this.state.elapsedMs;
    return this.state.elapsedMs + (nowMs2 - this.state.startAtMs);
  }
  snapshot(nowMs2 = performance.now()) {
    return {
      score: this.getScore(),
      elapsedMs: Math.max(0, Math.floor(this.getElapsedMs(nowMs2))),
      running: this.state.running
    };
  }
  static formatMs(ms) {
    const totalSec = Math.max(0, Math.floor(ms / 1e3));
    const m = Math.floor(totalSec / 60);
    const s = totalSec % 60;
    const mm = String(m).padStart(2, "0");
    const ss = String(s).padStart(2, "0");
    return `${mm}:${ss}`;
  }
};

// src/core/runtime/ui/hudDom.ts
var ScoreTimerHudDom = class {
  constructor(opts) {
    __publicField(this, "root");
    __publicField(this, "scoreEl");
    __publicField(this, "timeEl");
    const mount = opts?.mount ?? document.body;
    const safeTop = opts?.safeTop ?? 0;
    const safeLeft = opts?.safeLeft ?? 0;
    const zIndex = opts?.zIndex ?? 9999;
    const root = document.createElement("div");
    root.style.position = "fixed";
    root.style.top = `${safeTop + 12}px`;
    root.style.left = `${safeLeft + 12}px`;
    root.style.zIndex = String(zIndex);
    root.style.display = "flex";
    root.style.gap = "10px";
    root.style.alignItems = "center";
    root.style.padding = "10px 12px";
    root.style.borderRadius = "14px";
    root.style.background = "rgba(0,0,0,0.55)";
    root.style.backdropFilter = "blur(6px)";
    root.style.color = "#fff";
    root.style.fontFamily = "system-ui, -apple-system, Segoe UI, Roboto, Arial";
    root.style.fontSize = "14px";
    root.style.userSelect = "none";
    root.style.pointerEvents = "none";
    const score = document.createElement("span");
    score.textContent = "Score: 0";
    score.style.fontWeight = "700";
    const time = document.createElement("span");
    time.textContent = "Time: 00:00";
    time.style.opacity = "0.9";
    root.append(score, time);
    mount.appendChild(root);
    this.root = root;
    this.scoreEl = score;
    this.timeEl = time;
  }
  update(snapshot) {
    this.scoreEl.textContent = `Score: ${snapshot.score}`;
    this.timeEl.textContent = `Time: ${ScoreTimer.formatMs(snapshot.elapsedMs)}`;
  }
  setVisible(v) {
    this.root.style.display = v ? "flex" : "none";
  }
  destroy() {
    this.root.remove();
  }
};

// src/core/runtime/ui/resultOverlayDom.ts
var ResultOverlayDom = class {
  constructor(opts) {
    __publicField(this, "root");
    // private box: HTMLDivElement;
    __publicField(this, "scoreEl");
    __publicField(this, "timeEl");
    const mount = opts?.mount ?? document.body;
    const zIndex = opts?.zIndex ?? 1e4;
    const safeBottom = opts?.safeBottom ?? 0;
    const root = document.createElement("div");
    root.style.position = "fixed";
    root.style.inset = "0";
    root.style.zIndex = String(zIndex);
    root.style.display = "none";
    root.style.alignItems = "center";
    root.style.justifyContent = "center";
    root.style.background = "rgba(0,0,0,0.55)";
    root.style.backdropFilter = "blur(6px)";
    root.style.color = "#fff";
    root.style.fontFamily = "system-ui, -apple-system, Segoe UI, Roboto, Arial";
    const box = document.createElement("div");
    box.style.width = "min(520px, calc(100% - 32px))";
    box.style.borderRadius = "18px";
    box.style.background = "rgba(10, 16, 30, 0.95)";
    box.style.border = "1px solid rgba(255,255,255,0.12)";
    box.style.padding = "18px";
    box.style.boxShadow = "0 20px 60px rgba(0,0,0,0.35)";
    box.style.transform = `translateY(${-safeBottom / 2}px)`;
    const title = document.createElement("div");
    title.textContent = "K\u1EBFt qu\u1EA3";
    title.style.fontSize = "18px";
    title.style.fontWeight = "800";
    title.style.marginBottom = "12px";
    const scoreEl = document.createElement("div");
    scoreEl.style.fontSize = "16px";
    scoreEl.style.fontWeight = "700";
    const timeEl = document.createElement("div");
    timeEl.style.marginTop = "6px";
    timeEl.style.opacity = "0.9";
    const btnRow = document.createElement("div");
    btnRow.style.display = "flex";
    btnRow.style.gap = "10px";
    btnRow.style.marginTop = "16px";
    const mkBtn = (label) => {
      const b = document.createElement("button");
      b.textContent = label;
      b.style.flex = "1";
      b.style.padding = "10px 12px";
      b.style.borderRadius = "12px";
      b.style.border = "0";
      b.style.cursor = "pointer";
      b.style.fontWeight = "700";
      return b;
    };
    const replay = mkBtn("Ch\u01A1i l\u1EA1i");
    replay.style.background = "#2b5cff";
    replay.style.color = "#fff";
    replay.onclick = () => opts?.onReplay?.();
    const exit = mkBtn("Tho\xE1t");
    exit.style.background = "rgba(255,255,255,0.10)";
    exit.style.color = "#fff";
    exit.onclick = () => opts?.onExit?.();
    btnRow.append(replay, exit);
    box.append(title, scoreEl, timeEl, btnRow);
    root.appendChild(box);
    mount.appendChild(root);
    this.root = root;
    this.scoreEl = scoreEl;
    this.timeEl = timeEl;
  }
  show(data) {
    this.scoreEl.textContent = `Score: ${data.score}`;
    this.timeEl.textContent = `Time: ${ScoreTimer.formatMs(data.elapsedMs)}`;
    this.root.style.display = "flex";
  }
  hide() {
    this.root.style.display = "none";
  }
  destroy() {
    this.root.remove();
  }
};

// src/core/runtime/gameSessionUiRuntime.ts
function createGameSessionUiRuntime(opts) {
  const timer = new ScoreTimer();
  const hud = new ScoreTimerHudDom({
    safeTop: opts.safeInsets?.top ?? 0,
    safeLeft: opts.safeInsets?.left ?? 0
  });
  let completed = false;
  const overlay = new ResultOverlayDom({
    safeBottom: opts.safeInsets?.bottom ?? 0,
    onReplay: () => opts.onReplay?.(),
    onExit: () => opts.onExit?.()
  });
  let raf = null;
  const tick = () => {
    const snap = timer.snapshot();
    hud.update(snap);
    raf = requestAnimationFrame(tick);
  };
  const start = () => {
    timer.start();
    if (raf == null) raf = requestAnimationFrame(tick);
  };
  const pause = () => timer.pause();
  const resume = () => timer.start();
  const setScore = (score, delta) => {
    timer.setScore(score);
    opts.sdk.score(score, delta);
  };
  const addScore = (delta) => {
    timer.addScore(delta);
    opts.sdk.score(timer.getScore(), delta);
  };
  const finish = (extras) => {
    if (completed) return;
    completed = true;
    const snap = timer.snapshot();
    overlay.show(snap);
    opts.sdk.complete({
      score: snap.score,
      timeMs: snap.elapsedMs,
      extras
    });
  };
  const destroy = () => {
    if (raf != null) cancelAnimationFrame(raf);
    raf = null;
    hud.destroy();
    overlay.destroy();
  };
  return {
    timer,
    hud,
    overlay,
    start,
    pause,
    resume,
    setScore,
    addScore,
    finish,
    destroy
  };
}

// src/core/runtime/state/stateClient.ts
var StateClient = class {
  constructor(sdk) {
    __publicField(this, "sdk");
    __publicField(this, "pendingLoad", null);
    this.sdk = sdk;
  }
  /** Game xin Hub lưu state */
  save(state2) {
    this.sdk.requestSave(state2);
  }
  /** Game xin Hub trả state */
  load() {
    if (this.pendingLoad) {
      return Promise.reject(new Error("LOAD_ALREADY_PENDING"));
    }
    return new Promise((resolve) => {
      this.pendingLoad = resolve;
      this.sdk.requestLoad();
    });
  }
  /** Hub gọi SET_STATE → game nhận ở đây */
  handleSetState(state2) {
    if (this.pendingLoad) {
      this.pendingLoad(state2);
      this.pendingLoad = null;
    }
  }
};

// src/core/phaser/index.ts
var phaser_exports = {};
__export(phaser_exports, {
  AssetManager: () => AssetManager,
  HowlerAudioManager: () => HowlerAudioManager,
  ScaleManager: () => ScaleManager,
  createEndGameScene: () => createEndGameScene,
  createScaleManager: () => createScaleManager
});

// src/core/phaser/scaleManager.ts
var ScaleManager = class {
  constructor(scene, options) {
    __publicField(this, "scene");
    __publicField(this, "opts");
    __publicField(this, "graphics");
    __publicField(this, "lastSnap");
    this.scene = scene;
    this.opts = {
      design: options.design,
      maxDpr: options.maxDpr ?? 2,
      safeAreaMode: options.safeAreaMode ?? "css-vars",
      debug: options.debug ?? false,
      onChange: options.onChange
    };
    if (this.opts.debug) {
      this.graphics = scene.add.graphics().setDepth(999999);
    }
  }
  /** gọi 1 lần trong create() */
  initFromWindow() {
    const w = window.innerWidth;
    const h = window.innerHeight;
    const dpr = window.devicePixelRatio || 1;
    this.applyResize({ width: w, height: h, dpr });
  }
  /** gọi khi hub gửi RESIZE */
  applyResize(input) {
    const dpr = clampDpr(
      input.dpr ?? window.devicePixelRatio ?? 1,
      this.opts.maxDpr
    );
    const game = this.scene.game;
    if (game.renderer && "resolution" in game.renderer) {
      game.renderer.resolution = dpr;
    }
    this.scene.scale.resize(input.width, input.height);
    this.scene.scale.refresh();
    const safeArea = this.opts.safeAreaMode === "css-vars" ? readSafeAreaCssVars() : zeroInsets();
    const snap = computeFitSnapshot({
      container: { width: input.width, height: input.height, dpr },
      design: this.opts.design,
      safeArea
    });
    this.applyCameraFit(snap);
    if (!this.lastSnap || !snapEquals(this.lastSnap, snap)) {
      this.lastSnap = snap;
      if (this.opts.debug) this.drawDebug(snap);
      this.opts.onChange?.(snap);
      this.scene.events.emit("scale:snapshot", snap);
    }
  }
  /** snapshot mới nhất */
  getSnapshot() {
    if (!this.lastSnap) {
      return computeFitSnapshot({
        container: {
          width: this.scene.scale.width,
          height: this.scene.scale.height,
          dpr: clampDpr(window.devicePixelRatio || 1, this.opts.maxDpr)
        },
        design: this.opts.design,
        safeArea: this.opts.safeAreaMode === "css-vars" ? readSafeAreaCssVars() : zeroInsets()
      });
    }
    return this.lastSnap;
  }
  /** tiện: trả viewport rect (nơi nên đặt UI) */
  getViewport() {
    return this.getSnapshot().viewport;
  }
  /** tiện: trả safeArea */
  getSafeArea() {
    return this.getSnapshot().safeArea;
  }
  /**
   * Quy ước:
   * - World size = design (0..designW, 0..designH)
   * - Camera zoom = snap.scale
   * - Camera scroll để world center vào viewport
   */
  applyCameraFit(snap) {
    const cam = this.scene.cameras.main;
    cam.setViewport(
      snap.viewport.x,
      snap.viewport.y,
      snap.viewport.width,
      snap.viewport.height
    );
    cam.setBounds(0, 0, snap.design.width, snap.design.height);
    cam.setZoom(snap.scale);
    cam.centerOn(snap.design.width / 2, snap.design.height / 2);
  }
  drawDebug(snap) {
    if (!this.graphics) return;
    const g = this.graphics;
    g.clear();
    const safeX = snap.safeArea.left;
    const safeY = snap.safeArea.top;
    const safeW = snap.container.width - snap.safeArea.left - snap.safeArea.right;
    const safeH = snap.container.height - snap.safeArea.top - snap.safeArea.bottom;
    g.lineStyle(2, 65280, 1);
    g.strokeRect(safeX, safeY, safeW, safeH);
    g.lineStyle(2, 16711935, 1);
    g.strokeRect(
      snap.viewport.x,
      snap.viewport.y,
      snap.viewport.width,
      snap.viewport.height
    );
  }
};
function createScaleManager(scene, options) {
  return new ScaleManager(scene, options);
}
function clampDpr(dpr, maxDpr) {
  const n = Number.isFinite(dpr) ? dpr : 1;
  return Math.max(1, Math.min(maxDpr, n));
}
function zeroInsets() {
  return { top: 0, right: 0, bottom: 0, left: 0 };
}
function readSafeAreaCssVars() {
  const s = getComputedStyle(document.documentElement);
  const top = parseCssPx(s.getPropertyValue("--safe-area-inset-top"));
  const right = parseCssPx(s.getPropertyValue("--safe-area-inset-right"));
  const bottom = parseCssPx(s.getPropertyValue("--safe-area-inset-bottom"));
  const left = parseCssPx(s.getPropertyValue("--safe-area-inset-left"));
  return { top, right, bottom, left };
}
function parseCssPx(v) {
  const n = parseFloat(String(v).trim().replace("px", ""));
  return Number.isFinite(n) ? n : 0;
}
function computeFitSnapshot(args) {
  const { container, design, safeArea } = args;
  const safeW = Math.max(1, container.width - safeArea.left - safeArea.right);
  const safeH = Math.max(1, container.height - safeArea.top - safeArea.bottom);
  const scale = Math.min(safeW / design.width, safeH / design.height);
  const viewW = Math.round(design.width * scale);
  const viewH = Math.round(design.height * scale);
  const x = Math.round(safeArea.left + (safeW - viewW) / 2);
  const y = Math.round(safeArea.top + (safeH - viewH) / 2);
  return {
    container,
    safeArea,
    design,
    scale,
    viewport: { x, y, width: viewW, height: viewH }
  };
}
function snapEquals(a, b) {
  return a.container.width === b.container.width && a.container.height === b.container.height && a.safeArea.top === b.safeArea.top && a.safeArea.right === b.safeArea.right && a.safeArea.bottom === b.safeArea.bottom && a.safeArea.left === b.safeArea.left && Math.abs(a.scale - b.scale) < 1e-6 && a.viewport.x === b.viewport.x && a.viewport.y === b.viewport.y && a.viewport.width === b.viewport.width && a.viewport.height === b.viewport.height;
}

// src/core/phaser/assetManager.ts
import Phaser2 from "phaser";
var AssetManager = class {
  constructor(scene, opts) {
    __publicField(this, "scene");
    __publicField(this, "opts");
    this.scene = scene;
    this.opts = {
      preferNetworkProbe: opts?.preferNetworkProbe ?? true,
      probeTimeoutMs: opts?.probeTimeoutMs ?? 1500,
      debug: opts?.debug ?? false,
      strict: opts?.strict ?? false,
      onProgress: opts?.onProgress
    };
  }
  /**
   * Load một pack và đảm bảo decode image (phần decode là “giảm giật”)
   */
  async loadPack(pack) {
    const images = pack.images ?? [];
    const atlases = pack.atlases ?? [];
    const audio = pack.audio ?? [];
    const json = pack.json ?? [];
    const resolvedImages = await Promise.all(
      images.map((it) => this.resolveImage(it))
    );
    const resolvedAtlases = await Promise.all(
      atlases.map((it) => this.resolveAtlas(it))
    );
    const loader = this.scene.load;
    for (const img of resolvedImages) {
      if (!this.scene.textures.exists(img.key)) {
        loader.image(img.key, img.url);
        if (img.crossOrigin != null) loader.setCORS(img.crossOrigin);
      }
    }
    for (const at of resolvedAtlases) {
      if (!this.scene.textures.exists(at.key)) {
        loader.atlas(at.key, at.imageUrl, at.atlasJsonPath);
        if (at.crossOrigin != null) loader.setCORS(at.crossOrigin);
      }
    }
    for (const a of audio) {
      if (!this.scene.cache.audio.exists(a.key)) {
        loader.audio(a.key, a.paths);
      }
    }
    for (const j of json) {
      if (!this.scene.cache.json.exists(j.key)) {
        loader.json(j.key, j.path);
      }
    }
    const total = resolvedImages.length + resolvedAtlases.length + audio.length;
    if (total === 0) return;
    let loadedCount = 0;
    const emit = (currentKey, progress01) => {
      const p01 = progress01 ?? (total ? loadedCount / total : 1);
      this.opts.onProgress?.({
        loaded: loadedCount,
        total,
        progress01: clamp013(p01),
        currentKey
      });
    };
    loader.on(Phaser2.Loader.Events.FILE_COMPLETE, (key) => {
      loadedCount += 1;
      emit(key);
    });
    await new Promise((resolve, reject) => {
      let firstError = null;
      const onFileError = (file) => {
        firstError ?? (firstError = file);
      };
      loader.once(Phaser2.Loader.Events.FILE_LOAD_ERROR, onFileError);
      loader.once(
        Phaser2.Loader.Events.COMPLETE,
        async (_ldr, totalComplete, totalFailed) => {
          try {
            if (totalFailed > 0) {
              const f = firstError;
              reject(
                new Error(
                  `ASSET_LOAD_FAIL: failed=${totalFailed}, complete=${totalComplete}, first=${f?.key || "unknown"} url=${f?.src || f?.url || ""}`
                )
              );
              return;
            }
            await this.decodeTextures([
              ...resolvedImages.map((i) => i.key),
              ...resolvedAtlases.map((a) => a.key)
            ]);
            emit(void 0, 1);
            resolve();
          } catch (e) {
            reject(e);
          }
        }
      );
      emit(void 0, 0);
      loader.start();
    });
    loader.off(Phaser2.Loader.Events.FILE_LOAD_ERROR);
    loader.off(Phaser2.Loader.Events.FILE_COMPLETE);
  }
  // ----------------- resolve URLs -----------------
  async resolveImage(it) {
    const exts = it.exts ?? ["avif", "webp", "png", "jpg", "jpeg"];
    const url = await this.pickBest(`${it.basePath}`, exts);
    return { key: it.key, url, crossOrigin: it.crossOrigin };
  }
  async resolveAtlas(it) {
    const exts = it.exts ?? ["avif", "webp", "png"];
    const imageUrl = await this.pickBest(`${it.imageBasePath}`, exts);
    return {
      key: it.key,
      imageUrl,
      atlasJsonPath: it.atlasJsonPath,
      crossOrigin: it.crossOrigin
    };
  }
  /**
   * chọn extension theo:
   * - nếu preferNetworkProbe: HEAD lần lượt -> cái nào tồn tại dùng
   * - fallback: chọn theo support (avif/webp) rồi fallback png/jpg
   */
  async pickBest(base, exts) {
    const ordered = orderBySupport(exts);
    const tried = [];
    if (this.opts.preferNetworkProbe) {
      for (const ext of ordered) {
        const url = `${base}.${ext}`;
        tried.push(url);
        const ok2 = await headOk(url, this.opts.probeTimeoutMs);
        if (ok2) return url;
      }
      if (this.opts.strict) {
        throw new Error(`ASSET_MISSING: ${base} (tried: ${tried.join(", ")})`);
      }
      return `${base}.${ordered[0]}`;
    }
    return `${base}.${ordered[0]}`;
  }
  // ----------------- decode -----------------
  /**
   * Decode texture sources để giảm giật:
   * - tạo ImageBitmap nếu browser support (nhanh)
   * - hoặc fallback: tạo <img>.decode() qua network cache (chỉ làm được khi có URL)
   *
   * Lưu ý: Phaser texture có thể chứa nhiều source (atlas)
   */
  async decodeTextures(textureKeys) {
    const unique = Array.from(new Set(textureKeys));
    const tasks = [];
    for (const key of unique) {
      const tex = this.scene.textures.get(key);
      if (!tex) continue;
      const sources = tex.source ?? [];
      for (const s of sources) {
        const img = s?.image;
        if (!img) continue;
        if (img.__irukaDecoded) continue;
        tasks.push(
          decodeOne(img).then(() => {
            img.__irukaDecoded = true;
          })
        );
      }
    }
    await Promise.all(tasks);
  }
};
function clamp013(n) {
  return Math.max(0, Math.min(1, n));
}
function orderBySupport(exts) {
  const avifOk = supportsAvif();
  const webpOk = supportsWebp();
  const score = (ext) => {
    const e = ext.toLowerCase();
    if (e === "avif") return avifOk ? 0 : 100;
    if (e === "webp") return webpOk ? 1 : 101;
    if (e === "png") return 10;
    if (e === "jpg" || e === "jpeg") return 11;
    return 50;
  };
  return [...exts].sort((a, b) => score(a) - score(b));
}
function supportsWebp() {
  try {
    const c = document.createElement("canvas");
    return c.toDataURL("image/webp").startsWith("data:image/webp");
  } catch {
    return false;
  }
}
function supportsAvif() {
  try {
    return typeof CSS !== "undefined" && CSS.supports?.("image-rendering", "auto") ? true : false;
  } catch {
    return false;
  }
}
async function headOk(url, timeoutMs) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      method: "HEAD",
      mode: "cors",
      cache: "no-cache",
      signal: ctrl.signal
    });
    return res.ok;
  } catch {
    return false;
  } finally {
    clearTimeout(t);
  }
}
async function decodeOne(img) {
  if (typeof img.decode === "function") {
    try {
      await img.decode();
    } catch {
    }
    return;
  }
  return;
}

// src/core/phaser/HowlerAudioManager.ts
import { Howl, Howler } from "howler";
var HowlerAudioManager = class {
  constructor(opts) {
    __publicField(this, "tracks", /* @__PURE__ */ new Map());
    __publicField(this, "unlocked", false);
    __publicField(this, "muted", false);
    __publicField(this, "bgmKey", null);
    __publicField(this, "voiceKey", null);
    __publicField(this, "pendingPlays", []);
    __publicField(this, "queueUntilUnlock", true);
    __publicField(this, "opts");
    this.opts = {
      groupVolume: {
        bgm: opts?.groupVolume?.bgm ?? 1,
        voice: opts?.groupVolume?.voice ?? 1,
        sfx: opts?.groupVolume?.sfx ?? 1
      },
      duckBgmTo: opts?.duckBgmTo ?? 0.25,
      fadeMs: opts?.fadeMs ?? 200,
      debug: opts?.debug ?? false,
      onState: opts?.onState,
      defaultProbe: opts?.defaultProbe ?? true,
      defaultProbeTimeoutMs: opts?.defaultProbeTimeoutMs ?? 1200
    };
  }
  // ---------- registry ----------
  /**
   * Register assets. Supports:
   * - { src: [...] }
   * - { basePath, exts }  (auto build + optional probe)
   */
  async register(assets) {
    for (const a of assets) {
      if (this.tracks.has(a.key)) continue;
      const src = await this.resolveSrc(a);
      if (!src.length) {
        this.d("[audio] no valid src resolved for", a.key);
        continue;
      }
      const crossOrigin = a.crossOrigin;
      if (crossOrigin) Howler.autoUnlock = true;
      const howl = new Howl({
        src,
        preload: a.preload ?? true,
        loop: a.loop ?? false,
        volume: this.computeTrackVolume(a),
        rate: a.rate ?? 1,
        html5: a.kind === "bgm"
      });
      this.tracks.set(a.key, { asset: a, howl });
    }
  }
  has(key) {
    return this.tracks.has(key);
  }
  async loadAll(opts) {
    const strict = opts?.strict ?? false;
    const entries = [...this.tracks.entries()];
    const total = entries.length;
    let loaded = 0;
    const results = await Promise.all(
      entries.map(([key, t]) => {
        return new Promise(
          (resolve) => {
            const done = (ok2, err2) => {
              loaded += 1;
              opts?.onProgress?.({ loaded, total });
              resolve({ key, ok: ok2, err: err2 });
            };
            if (t.howl.state() === "loaded") return done(true);
            t.howl.once("load", () => done(true));
            t.howl.once("loaderror", (_id, err2) => done(false, err2));
            try {
              t.howl.load();
            } catch (e) {
              done(false, e);
            }
          }
        );
      })
    );
    const failed = results.filter((r) => !r.ok);
    if (strict && failed.length) {
      throw new Error(
        `AUDIO_LOAD_FAIL: ${failed.map((f) => f.key).join(", ")}`
      );
    }
  }
  // ---------- unlock ----------
  async unlock() {
    if (this.unlocked) return true;
    try {
      const ctx = Howler.ctx;
      if (ctx && ctx.state === "suspended") await ctx.resume();
      const silent = new Howl({
        src: [
          "data:audio/mp3;base64,//uQxAAAAAAAAAAAAAAAAAAAAAAAWGluZwAAAA8AAAACAAACcQCA"
        ],
        volume: 0
      });
      silent.play();
      silent.stop();
      this.unlocked = true;
      this.emitState();
      this.d("[audio] unlocked");
      const plays = [...this.pendingPlays];
      this.pendingPlays = [];
      plays.forEach((fn) => {
        try {
          fn();
        } catch {
        }
      });
      return true;
    } catch (e) {
      this.d("[audio] unlock failed", e);
      return false;
    }
  }
  bindUnlockToGesture(target = document) {
    const handler = async () => {
      await this.unlock();
      target.removeEventListener("pointerdown", handler);
      target.removeEventListener("touchstart", handler);
      target.removeEventListener("click", handler);
    };
    target.addEventListener("pointerdown", handler, {
      passive: true,
      once: true
    });
    target.addEventListener("touchstart", handler, {
      passive: true,
      once: true
    });
    target.addEventListener("click", handler, {
      passive: true,
      once: true
    });
  }
  // ---------- controls ----------
  setMuted(muted) {
    this.muted = muted;
    Howler.mute(muted);
    this.emitState();
  }
  isUnlocked() {
    return this.unlocked;
  }
  // ---------- play APIs ----------
  playSfx(key) {
    return this.play(key, "sfx");
  }
  playVoice(key) {
    return this.play(key, "voice");
  }
  async setBgm(key) {
    if (this.bgmKey === key) {
      const t2 = this.tracks.get(key);
      if (t2 && !t2.howl.playing()) t2.howl.play();
      return;
    }
    const prev = this.bgmKey;
    this.bgmKey = key;
    if (prev) this.fadeOutStop(prev);
    const t = this.tracks.get(key);
    if (!t) return;
    t.howl.volume(0);
    t.howl.loop(true);
    t.howl.play();
    this.fadeTo(key, this.computeTrackVolume(t.asset), this.opts.fadeMs);
  }
  stopBgm() {
    if (!this.bgmKey) return;
    this.fadeOutStop(this.bgmKey);
    this.bgmKey = null;
  }
  stopAllVoices() {
    for (const [_key, t] of this.tracks.entries()) {
      if (t.asset.kind === "voice") {
        try {
          t.howl.stop();
        } catch {
        }
      }
    }
    if (this.voiceKey) this.voiceKey = null;
    this.duckBgm(false);
  }
  stopAll() {
    for (const [key] of this.tracks) this.stop(key);
    this.bgmKey = null;
    this.voiceKey = null;
  }
  pauseAll() {
    for (const t of this.tracks.values()) {
      if (t.howl.playing()) t.howl.pause();
    }
  }
  resumeAll() {
    if (this.bgmKey) {
      const t = this.tracks.get(this.bgmKey);
      if (t && !t.howl.playing()) t.howl.play();
    }
    if (this.voiceKey) {
      const t = this.tracks.get(this.voiceKey);
      if (t && !t.howl.playing()) t.howl.play();
    }
  }
  // ---------- internals ----------
  async resolveSrc(a) {
    if (a.src?.length) return a.src;
    if (!a.basePath) return [];
    const exts = a.exts ?? ["m4a", "mp3", "ogg"];
    const urls = exts.map((ext) => `${a.basePath}.${ext}`);
    const probe = a.probe ?? this.opts.defaultProbe;
    if (!probe) return urls;
    const timeout = a.probeTimeoutMs ?? this.opts.defaultProbeTimeoutMs;
    const existed = await filterExistingUrls(urls, timeout);
    return existed;
  }
  play(key, kind) {
    const t = this.tracks.get(key);
    if (!t) {
      this.d("[audio] missing:", key);
      return null;
    }
    const doPlay = () => {
      if (kind === "voice") {
        if (this.voiceKey && this.voiceKey !== key) this.stop(this.voiceKey);
        this.voiceKey = key;
        this.duckBgm(true);
      }
      t.howl.volume(this.computeTrackVolume(t.asset));
      const id = t.howl.play();
      t.howl.once(
        "end",
        () => {
          if (kind === "voice" && this.voiceKey === key) {
            this.voiceKey = null;
            this.duckBgm(false);
          }
        },
        id
      );
      t.howl.once(
        "stop",
        () => {
          if (kind === "voice" && this.voiceKey === key) {
            this.voiceKey = null;
            this.duckBgm(false);
          }
        },
        id
      );
      return id;
    };
    if (!this.unlocked && this.queueUntilUnlock) {
      this.pendingPlays.push(doPlay);
      this.d("[audio] queue play until unlock:", key);
      return null;
    }
    return doPlay();
  }
  stop(key) {
    const t = this.tracks.get(key);
    if (!t) return;
    try {
      t.howl.stop();
    } catch {
    }
  }
  fadeOutStop(key) {
    const t = this.tracks.get(key);
    if (!t) return;
    try {
      const from = t.howl.volume();
      t.howl.fade(from, 0, this.opts.fadeMs);
      window.setTimeout(() => {
        try {
          t.howl.stop();
        } catch {
        }
        t.howl.volume(this.computeTrackVolume(t.asset));
      }, this.opts.fadeMs + 30);
    } catch {
      this.stop(key);
    }
  }
  fadeTo(key, to, ms) {
    const t = this.tracks.get(key);
    if (!t) return;
    try {
      const from = t.howl.volume();
      t.howl.fade(from, to, ms);
    } catch {
      t.howl.volume(to);
    }
  }
  duckBgm(duck) {
    if (!this.bgmKey) return;
    const t = this.tracks.get(this.bgmKey);
    if (!t) return;
    const normal = this.computeTrackVolume(t.asset);
    const target = duck ? Math.min(normal, this.opts.duckBgmTo) : normal;
    this.fadeTo(this.bgmKey, target, this.opts.fadeMs);
  }
  computeTrackVolume(a) {
    const base = a.volume ?? 1;
    const g = this.opts.groupVolume[a.kind] ?? 1;
    return clamp014(base * g);
  }
  emitState() {
    this.opts.onState?.({ unlocked: this.unlocked, muted: this.muted });
  }
  d(...args) {
    if (this.opts.debug) console.log(...args);
  }
};
function clamp014(n) {
  return Math.max(0, Math.min(1, n));
}
async function filterExistingUrls(urls, timeoutMs) {
  const ok2 = [];
  for (const url of urls) {
    const exists = await headOk2(url, timeoutMs);
    if (exists) ok2.push(url);
  }
  return ok2.length ? ok2 : urls;
}
async function headOk2(url, timeoutMs) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      method: "HEAD",
      cache: "no-cache",
      signal: ctrl.signal
    });
    return res.ok;
  } catch {
    return false;
  } finally {
    clearTimeout(t);
  }
}

// src/core/phaser/createEndGameScene.ts
import Phaser3 from "phaser";
function createEndGameScene(options) {
  const KEY = options.sceneKey ?? "EndGameScene";
  const base = options.designBase ?? { w: 1280, h: 720 };
  const confettiColors = options.confetti?.colors ?? [
    16739179,
    5164484,
    16770669,
    9822675,
    15958401,
    11179738
  ];
  return class SharedEndGameScene extends Phaser3.Scene {
    constructor() {
      super(KEY);
      __publicField(this, "confettiEvent");
      __publicField(this, "dimBg");
      __publicField(this, "banner");
      __publicField(this, "icon");
      __publicField(this, "replayBtn");
      __publicField(this, "exitBtn");
      __publicField(this, "dataIn");
    }
    init(data) {
      this.dataIn = data;
    }
    preload() {
      const { assets } = options;
      const loadIfNeeded = (key, url) => {
        if (!url) return;
        if (this.textures.exists(key)) return;
        this.load.image(key, url);
      };
      loadIfNeeded(assets.banner.key, assets.banner.url);
      if (assets.icon) loadIfNeeded(assets.icon.key, assets.icon.url);
      loadIfNeeded(assets.replayBtn.key, assets.replayBtn.url);
      loadIfNeeded(assets.exitBtn.key, assets.exitBtn.url);
    }
    create() {
      options.onEnter?.();
      if (options.audio && options.sounds?.enter)
        options.audio.play(options.sounds.enter);
      this.time.delayedCall(2e3, () => {
        if (!options.audio) return;
        if (options.sounds?.fireworks)
          options.audio.play(options.sounds.fireworks);
        if (options.sounds?.applause)
          options.audio.play(options.sounds.applause);
      });
      this.buildUI();
      this.layout();
      this.scale.on("resize", () => this.layout());
      if (options.confetti?.enabled !== false) this.startConfetti();
    }
    buildUI() {
      const w = this.scale.width;
      const h = this.scale.height;
      this.banner = this.add.image(w / 2, h / 2 - h * 0.12, options.assets.banner.key).setOrigin(0.5).setDepth(100);
      if (options.assets.icon && this.textures.exists(options.assets.icon.key)) {
        this.icon = this.add.image(w / 2, h / 2 - 150, options.assets.icon.key).setDepth(110);
        this.icon.setScale(0.5);
        this.tweens.add({
          targets: this.icon,
          y: this.icon.y - 10,
          duration: 800,
          yoyo: true,
          repeat: -1,
          ease: "Sine.easeInOut"
        });
        this.tweens.add({
          targets: this.icon,
          angle: { from: -5, to: 5 },
          duration: 600,
          yoyo: true,
          repeat: -1,
          ease: "Sine.easeInOut"
        });
      }
      const btnScale = Math.min(w, h) / 1280;
      const spacing = 250 * btnScale;
      this.replayBtn = this.add.image(w / 2 - spacing, h / 2 + h * 0.2, options.assets.replayBtn.key).setOrigin(0.5).setDepth(120).setInteractive({ useHandCursor: true });
      this.exitBtn = this.add.image(w / 2 + spacing, h / 2 + h * 0.2, options.assets.exitBtn.key).setOrigin(0.5).setDepth(120).setInteractive({ useHandCursor: true });
      this.replayBtn.on("pointerdown", () => this.onReplay());
      this.exitBtn.on("pointerdown", () => this.onExit());
      [this.replayBtn, this.exitBtn].forEach((btn) => {
        btn.on("pointerover", () => btn.setScale(btn.scale * 1.1));
        btn.on("pointerout", () => this.layout());
      });
    }
    layout() {
      const w = this.scale.width;
      const h = this.scale.height;
      if (this.dimBg) this.dimBg.setSize(w, h);
      const scale = Math.min(w / base.w, h / base.h);
      if (this.banner) {
        this.banner.setPosition(w / 2, h / 2 - h * 0.12);
        this.banner.setDisplaySize(w * 0.9, h * 0.9);
      }
      if (this.icon) {
        this.icon.setPosition(w / 2, h / 2 - 150 * scale);
        this.icon.setScale(0.5 * scale);
      }
      const btnScale = Math.min(w, h) / 1280;
      const spacing = 250 * btnScale;
      if (this.replayBtn) {
        this.replayBtn.setPosition(w / 2 - spacing, h / 2 + h * 0.2);
        this.replayBtn.setScale(btnScale);
      }
      if (this.exitBtn) {
        this.exitBtn.setPosition(w / 2 + spacing, h / 2 + h * 0.2);
        this.exitBtn.setScale(btnScale);
      }
    }
    onReplay() {
      if (options.audio && options.sounds?.click)
        options.audio.play(options.sounds.click);
      options.audio?.stopAll?.();
      retryFromStart();
      this.stopConfetti();
      if (options.replaySceneKey) {
        options.onLeave?.();
        this.scene.stop(KEY);
        this.scene.start(options.replaySceneKey);
      } else {
        this.scene.restart();
      }
    }
    onExit() {
      if (options.audio && options.sounds?.click)
        options.audio.play(options.sounds.click);
      options.audio?.stopAll?.();
      this.stopConfetti();
      const startTime = this.dataIn?.startTime ?? Date.now();
      const timeMs = Math.max(0, Date.now() - startTime);
      const stats = prepareSubmitData();
      const payload = {
        // score: giữ theo contract hiện tại (Hub đang đọc score)
        score: stats.finalScore ?? this.dataIn?.score ?? 0,
        timeMs,
        extras: {
          ...this.dataIn?.extras ?? {},
          reason: "user_exit",
          stats
          // nhét toàn bộ payload chuẩn vào đây
        }
      };
      if (options.reportComplete) {
        options.reportComplete(payload);
        return;
      }
      if (options.fallbackSceneKey) {
        options.onLeave?.();
        this.scene.stop(KEY);
        this.scene.start(options.fallbackSceneKey);
      }
    }
    startConfetti() {
      const rateMs = options.confetti?.rateMs ?? 100;
      this.confettiEvent = this.time.addEvent({
        delay: rateMs,
        loop: true,
        callback: () => {
          if (!this.scene.isActive()) return;
          const width = this.cameras.main.width;
          for (let i = 0; i < 3; i++) {
            this.createConfettiPiece(
              Phaser3.Math.Between(0, width),
              -20,
              Phaser3.Utils.Array.GetRandom(confettiColors),
              Phaser3.Utils.Array.GetRandom(["circle", "rect"])
            );
          }
        }
      });
    }
    createConfettiPiece(x, y, color, shape) {
      let confetti;
      if (shape === "circle") {
        confetti = this.add.circle(x, y, Phaser3.Math.Between(4, 8), color, 1);
      } else {
        confetti = this.add.rectangle(
          x,
          y,
          Phaser3.Math.Between(6, 12),
          Phaser3.Math.Between(10, 20),
          color,
          1
        );
      }
      confetti.setDepth(999);
      confetti.setRotation(Phaser3.Math.Between(0, 360) * Math.PI / 180);
      const duration = Phaser3.Math.Between(3e3, 5e3);
      const targetY = this.cameras.main.height + 50;
      const drift = Phaser3.Math.Between(-100, 100);
      this.tweens.add({
        targets: confetti,
        y: targetY,
        x: x + drift,
        rotation: confetti.rotation + Phaser3.Math.Between(2, 4) * Math.PI,
        duration,
        ease: "Linear",
        onComplete: () => confetti.destroy()
      });
      this.tweens.add({
        targets: confetti,
        alpha: { from: 1, to: 0.3 },
        duration,
        ease: "Cubic.easeIn"
      });
    }
    stopConfetti() {
      if (this.confettiEvent) {
        this.confettiEvent.remove(false);
        this.confettiEvent = void 0;
      }
    }
  };
}

// src/state/AutoSaveManager.ts
var AutoSaveManager = class {
  /**
   * @param saveCallback - Function to persist state to Hub
   * @param debounceMs - Delay in ms before saving (default: 3000ms)
   */
  constructor(saveCallback, debounceMs = 3e3) {
    __publicField(this, "debounceMs");
    __publicField(this, "pendingState", null);
    __publicField(this, "saveTimer", null);
    __publicField(this, "saveCallback");
    __publicField(this, "isSaving", false);
    this.saveCallback = saveCallback;
    this.debounceMs = debounceMs;
  }
  /**
   * Request a save. Will be debounced.
   * @param state - The game state to save
   */
  requestSave(state2) {
    this.pendingState = state2;
    if (this.saveTimer) {
      clearTimeout(this.saveTimer);
    }
    this.saveTimer = setTimeout(() => {
      this.flush();
    }, this.debounceMs);
  }
  /**
   * Force save immediately if there is pending state.
   * Useful for "onPause" or "onQuit" events.
   */
  async flush() {
    if (this.saveTimer) {
      clearTimeout(this.saveTimer);
      this.saveTimer = null;
    }
    if (!this.pendingState) return;
    if (this.isSaving) {
      this.saveTimer = setTimeout(() => this.flush(), 500);
      return;
    }
    try {
      this.isSaving = true;
      const stateToSave = this.pendingState;
      this.pendingState = null;
      await this.saveCallback(stateToSave);
      console.debug("[AutoSaveManager] Saved successfully.");
    } catch (err2) {
      console.error("[AutoSaveManager] Save failed:", err2);
    } finally {
      this.isSaving = false;
      if (this.pendingState) {
        this.requestSave(this.pendingState);
      }
    }
  }
};

// src/index.ts
import * as core from "@iruka-edu/game-core";

// src/core/voice/index.ts
var voice_exports = {};
__export(voice_exports, {
  EndSession: () => EndSession,
  ExerciseType: () => ExerciseType,
  StartSession: () => StartSession,
  Submit: () => Submit,
  VoiceApiError: () => VoiceApiError,
  VoiceSessionClient: () => VoiceSessionClient,
  configureVoiceSessionClient: () => configureVoiceSessionClient,
  getVoiceSessionClient: () => getVoiceSessionClient
});

// src/core/voice/types.ts
var ExerciseType = /* @__PURE__ */ ((ExerciseType2) => {
  ExerciseType2["NURSERY_RHYME"] = "NURSERY_RHYME";
  ExerciseType2["COUNTING"] = "COUNTING";
  ExerciseType2["SPELLING"] = "SPELLING";
  return ExerciseType2;
})(ExerciseType || {});

// src/core/voice/client.ts
var DEFAULT_BASE_URL = "https://iruka-api-1037337851453.asia-southeast1.run.app";
var DEFAULT_API_VERSION = "v1";
var VoiceApiError = class extends Error {
  constructor(args) {
    super(args.message);
    __publicField(this, "status");
    __publicField(this, "endpoint");
    __publicField(this, "detail");
    this.name = "VoiceApiError";
    this.status = args.status;
    this.endpoint = args.endpoint;
    this.detail = args.detail;
  }
};
function withTimeoutSignal(timeoutMs) {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), timeoutMs);
  return { signal: controller.signal, cancel: () => clearTimeout(t) };
}
async function resolveHeaders(headers) {
  if (!headers) return {};
  if (typeof headers === "function") return await headers();
  return headers;
}
async function safeReadErrorDetail(res) {
  const contentType = res.headers.get("content-type") || "";
  try {
    if (contentType.includes("application/json")) {
      const j = await res.json();
      return { detail: typeof j.detail === "string" ? j.detail : JSON.stringify(j) };
    }
    const text = await res.text();
    return { rawText: text };
  } catch {
    return {};
  }
}
function resolveStartMeta(req, defaultMeta) {
  const ctx = getSdkContext();
  return {
    gameSessionId: req.gameSessionId ?? ctx.sessionId ?? defaultMeta?.gameSessionId,
    gameId: req.gameId ?? ctx.gameId ?? defaultMeta?.gameId,
    lessonId: req.lessonId ?? ctx.lessonId ?? defaultMeta?.lessonId,
    gameVersion: req.gameVersion ?? ctx.gameVersion ?? defaultMeta?.gameVersion
  };
}
var VoiceSessionClient = class {
  constructor(config = {}) {
    __publicField(this, "baseUrl");
    __publicField(this, "apiVersion");
    __publicField(this, "timeoutMs");
    __publicField(this, "defaultTestMode");
    __publicField(this, "headers");
    __publicField(this, "defaultMeta");
    /** lưu voice sessionId gần nhất */
    __publicField(this, "lastVoiceSessionId", null);
    this.baseUrl = config.baseUrl || DEFAULT_BASE_URL;
    this.apiVersion = config.apiVersion || DEFAULT_API_VERSION;
    this.timeoutMs = config.timeoutMs ?? 6e4;
    this.defaultTestMode = config.testMode ?? false;
    this.headers = config.headers;
    this.defaultMeta = config.defaultMeta;
  }
  /** để test/dev nếu muốn set thẳng */
  setLastVoiceSessionId(id) {
    this.lastVoiceSessionId = id;
  }
  getLastVoiceSessionId() {
    return this.lastVoiceSessionId;
  }
  async startSession(request = {}) {
    const testmode = request.testmode ?? this.defaultTestMode;
    const endpoint = `/api/${this.apiVersion}/voice-sessions/start?testmode=${testmode}`;
    const url = `${this.baseUrl}${endpoint}`;
    const meta = resolveStartMeta(request, this.defaultMeta);
    if (!meta.gameId || !meta.gameVersion) {
      if (testmode) {
        meta.gameId ?? (meta.gameId = "local-game");
        meta.gameVersion ?? (meta.gameVersion = "0.0.0");
      } else {
        throw new Error(
          `Voice startSession missing meta (gameId/gameVersion). Call configureSdkContext({ fallback: ... }) or run inside Hub INIT.`
        );
      }
    }
    const extraHeaders = await resolveHeaders(this.headers);
    const { signal, cancel } = withTimeoutSignal(this.timeoutMs);
    try {
      const res = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...extraHeaders
        },
        body: JSON.stringify({
          gameSessionId: meta.gameSessionId,
          gameId: meta.gameId,
          lessonId: meta.lessonId,
          gameVersion: meta.gameVersion
        }),
        signal
      });
      if (!res.ok) {
        const err2 = await safeReadErrorDetail(res);
        throw new VoiceApiError({
          status: res.status,
          endpoint,
          detail: err2.detail,
          message: `Start session failed (${res.status})${err2.detail ? `: ${err2.detail}` : ""}`
        });
      }
      const data = await res.json();
      this.lastVoiceSessionId = data.sessionId;
      return data;
    } finally {
      cancel();
    }
  }
  resolveVoiceSessionId(explicit) {
    const id = explicit ?? this.lastVoiceSessionId;
    if (!id) {
      throw new Error(
        `Missing voice sessionId. Call StartSession() first (or pass sessionId explicitly).`
      );
    }
    return id;
  }
  async submitAnswer(request) {
    const sessionId = this.resolveVoiceSessionId(request.sessionId);
    const testmode = request.testmode ?? this.defaultTestMode;
    const endpoint = `/api/${this.apiVersion}/voice-sessions/${sessionId}/submit?testmode=${testmode}`;
    const url = `${this.baseUrl}${endpoint}`;
    const extraHeaders = await resolveHeaders(this.headers);
    const { signal, cancel } = withTimeoutSignal(this.timeoutMs);
    try {
      const form = new FormData();
      const fileName = typeof request.audioFile?.name === "string" ? request.audioFile.name : "audio.wav";
      form.append("audio_file", request.audioFile, fileName);
      form.append("questionIndex", String(request.questionIndex));
      form.append(
        "targetText",
        typeof request.targetText === "string" ? request.targetText : JSON.stringify(request.targetText)
      );
      form.append("durationMs", String(request.durationMs));
      form.append("exercise_type", request.exerciseType);
      const res = await fetch(url, {
        method: "POST",
        headers: { ...extraHeaders },
        body: form,
        signal
      });
      if (!res.ok) {
        const err2 = await safeReadErrorDetail(res);
        throw new VoiceApiError({
          status: res.status,
          endpoint,
          detail: err2.detail,
          message: `Submit answer failed (${res.status})${err2.detail ? `: ${err2.detail}` : ""}`
        });
      }
      return await res.json();
    } finally {
      cancel();
    }
  }
  async endSession(request) {
    const sessionId = this.resolveVoiceSessionId(request.sessionId);
    const testmode = request.testmode ?? this.defaultTestMode;
    const endpoint = `/api/${this.apiVersion}/voice-sessions/${sessionId}/end?testmode=${testmode}`;
    const url = `${this.baseUrl}${endpoint}`;
    const extraHeaders = await resolveHeaders(this.headers);
    const { signal, cancel } = withTimeoutSignal(this.timeoutMs);
    try {
      const res = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          ...extraHeaders
        },
        body: JSON.stringify({
          sessionId,
          totalQuestionsExpect: request.totalQuestionsExpect,
          isUserAborted: request.isUserAborted
        }),
        signal
      });
      if (!res.ok) {
        const err2 = await safeReadErrorDetail(res);
        throw new VoiceApiError({
          status: res.status,
          endpoint,
          detail: err2.detail,
          message: `End session failed (${res.status})${err2.detail ? `: ${err2.detail}` : ""}`
        });
      }
      const data = await res.json();
      this.lastVoiceSessionId = null;
      return data;
    } finally {
      cancel();
    }
  }
};
var _defaultClient = new VoiceSessionClient();
function configureVoiceSessionClient(config) {
  _defaultClient = new VoiceSessionClient(config);
}
function getVoiceSessionClient() {
  return _defaultClient;
}
async function StartSession(req = {}) {
  return _defaultClient.startSession(req);
}
async function Submit(req) {
  return _defaultClient.submitAnswer(req);
}
async function EndSession(req) {
  return _defaultClient.endSession(req);
}

// src/index.ts
var SDK_VERSION = "__SDK_VERSION__";
export {
  AutoSaveManager,
  BaseGame,
  CONTRACT_VERSION,
  HandshakeHubBridge,
  IRUKA_PLUGIN_KEY,
  IrukaGameHost,
  IrukaPlugin,
  MAX_MESSAGE_BYTES2 as MAX_MESSAGE_BYTES,
  OverlayManager,
  SDK_VERSION,
  __testSpy,
  clamp015 as clamp01,
  configureSdkContext,
  connectToHub2 as connectToHub,
  core,
  createIrukaPluginConfig,
  estimateJsonBytes2 as estimateJsonBytes,
  game_exports as game,
  getSdkContext,
  isGameEventType,
  isHubCommandType2 as isHubCommandType,
  isRecord2 as isRecord,
  isValidEnvelope2 as isValidEnvelope,
  makeEnvelope2 as makeEnvelope,
  normalizeResult,
  phaser_exports as phaser,
  resetSdkContext,
  runtime_exports as runtime,
  setSdkContext,
  validateManifest,
  validateMessageSize2 as validateMessageSize,
  voice_exports as voice
};
