// @iruka-edu/game-core/dist/index.cjs
'use strict';

var zod = require('zod');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);

// src/contract/types.ts
var CONTRACT_VERSION = "0.2.1";

// src/contract/errors.ts
var ErrorCode = /* @__PURE__ */ ((ErrorCode2) => {
  ErrorCode2["SYS_INIT_TIMEOUT"] = "SYS:INIT_TIMEOUT";
  ErrorCode2["SYS_INVALID_CONTEXT"] = "SYS:INVALID_CONTEXT";
  ErrorCode2["SYS_CONNECTION_LOST"] = "SYS:CONNECTION_LOST";
  ErrorCode2["GAME_CRASH"] = "GAME:CRASH";
  ErrorCode2["GAME_ASSET_FAIL"] = "GAME:ASSET_FAIL";
  ErrorCode2["DATA_VALIDATION_FAIL"] = "DATA:INVALID";
  return ErrorCode2;
})(ErrorCode || {});
var SdkError = class extends Error {
  constructor(code, message, details) {
    super(message);
    this.code = code;
    this.details = details;
    this.name = "SdkError";
  }
};
var LaunchContextSchema = zod.z.object({
  // Bắt buộc phải có
  sessionId: zod.z.string().min(1),
  gameId: zod.z.string().min(1),
  // Optional Technical Fields
  sdkVersion: zod.z.string().optional(),
  lessonId: zod.z.string().optional(),
  gameVersion: zod.z.string().optional(),
  // Có default value (Phòng trường hợp BE quên gửi)
  mode: zod.z.enum(["production", "dev", "review"]).default("production"),
  locale: zod.z.string().default("vi"),
  difficulty: zod.z.enum(["easy", "medium", "hard"]).default("medium"),
  // Optional Objects
  player: zod.z.object({
    id: zod.z.string().optional(),
    name: zod.z.string().default("Student"),
    avatar: zod.z.string().optional()
  }).optional().default({}),
  // Default object rỗng để tránh lỗi null access
  // Settings
  soundEnabled: zod.z.boolean().default(true)
}).passthrough();
var GameResultSchema = zod.z.object({
  score: zod.z.number().min(0, "Score cannot be negative"),
  completion: zod.z.number().min(0).max(1, "Completion must be 0-1"),
  accuracy: zod.z.number().min(0).max(1).optional(),
  // Metadata mở rộng (cho phép gửi object tùy ý nhưng phải an toàn)
  details: zod.z.record(zod.z.unknown()).optional(),
  timestamp: zod.z.number().default(() => Date.now())
});

// src/protocol/payload.ts
var MessageType = {
  // System handshake
  HELLO: "SYS:HELLO",
  // Game -> Hub: "I'm online"
  INIT: "CMD:INIT",
  // Hub -> Game: "Here's your context"
  // Game lifecycle commands (Hub -> Game)
  START: "CMD:START",
  PAUSE: "CMD:PAUSE",
  RESUME: "CMD:RESUME",
  RESIZE: "CMD:RESIZE",
  QUIT: "CMD:QUIT",
  SET_STATE: "CMD:SET_STATE",
  // Game events (Game -> Hub)
  READY: "EVT:READY",
  LOADING: "EVT:LOADING",
  SCORE_UPDATE: "EVT:SCORE_UPDATE",
  PROGRESS: "EVT:PROGRESS",
  COMPLETE: "EVT:COMPLETE",
  ERROR: "EVT:ERROR",
  REQUEST_SAVE: "EVT:REQUEST_SAVE",
  REQUEST_LOAD: "EVT:REQUEST_LOAD"
};
function makeEnvelope(params) {
  return {
    sdkVersion: params.sdkVersion,
    sessionId: params.sessionId,
    gameId: params.gameId,
    lessonId: params.lessonId,
    gameVersion: params.gameVersion,
    source: params.source,
    type: params.type,
    payload: params.payload
  };
}
function isRecord(v) {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}
function isValidEnvelope(v) {
  if (!isRecord(v)) return false;
  if (typeof v.sdkVersion !== "string") return false;
  if (typeof v.sessionId !== "string") return false;
  if (typeof v.gameId !== "string") return false;
  if (v.source !== "hub" && v.source !== "game") return false;
  if (typeof v.type !== "string") return false;
  return true;
}
var HUB_COMMAND_TYPES = [
  "INIT",
  "START",
  "PAUSE",
  "RESUME",
  "RESIZE",
  "QUIT",
  "SET_STATE"
];
var GAME_EVENT_TYPES = [
  "READY",
  "LOADING",
  "SCORE_UPDATE",
  "PROGRESS",
  "COMPLETE",
  "ERROR",
  "REQUEST_SAVE",
  "REQUEST_LOAD"
];
function isHubCommandType(t) {
  return HUB_COMMAND_TYPES.includes(t);
}
function isGameEventType(t) {
  return GAME_EVENT_TYPES.includes(t);
}
function clamp01(n) {
  if (!Number.isFinite(n)) return 0;
  return Math.min(1, Math.max(0, n));
}
var MAX_MESSAGE_BYTES = 16 * 1024;
function estimateJsonBytes(value) {
  try {
    return new TextEncoder().encode(JSON.stringify(value)).length;
  } catch {
    return 0;
  }
}
function validateMessageSize(envelope) {
  const size = estimateJsonBytes(envelope);
  if (size > MAX_MESSAGE_BYTES) {
    return {
      ok: false,
      error: `Message too large: ${size} bytes (max ${MAX_MESSAGE_BYTES})`
    };
  }
  return { ok: true };
}

// src/bridge/iframeBridge.ts
var MOCK_CONTEXT = {
  sessionId: "dev-session-local",
  gameId: "local-dev-game",
  mode: "dev",
  locale: "vi",
  difficulty: "medium",
  player: { name: "Developer", id: "dev-001" },
  soundEnabled: true
};
function connectToHub(targetOrigin = "*", timeoutMs = 2e3) {
  return new Promise((resolve, reject) => {
    let handshakeInterval = null;
    let isConnected = false;
    const listeners = /* @__PURE__ */ new Map();
    const sendMessage = (type, payload) => {
      if (type === MessageType.COMPLETE && payload) {
        const parse = GameResultSchema.safeParse(payload);
        if (!parse.success) {
          console.warn("\u26A0\uFE0F Game \u0111ang g\u1EEDi k\u1EBFt qu\u1EA3 kh\xF4ng chu\u1EA9n:", parse.error);
          return;
        }
        payload = parse.data;
      }
      if (isConnected) {
        const envelope = {
          source: "game",
          type,
          payload,
          sdkVersion: "1.0.0",
          sessionId: "pending",
          gameId: "unknown"
        };
        window.parent.postMessage(envelope, targetOrigin);
      } else {
        console.debug(
          `%c[MockBridge \u{1F4E4}] ${type}`,
          "color: green; font-weight: bold",
          payload
        );
      }
    };
    const messageHandler = (event) => {
      const data = event.data;
      if (data?.source !== "hub") return;
      if (data.type === MessageType.INIT && !isConnected) {
        try {
          const validContext = LaunchContextSchema.parse(data.payload);
          console.log("\u2705 [GameSDK] Connected to Real Hub.");
          clearInterval(handshakeInterval);
          isConnected = true;
          const bridge = {
            context: validContext,
            send: sendMessage,
            on: (type, cb) => {
              window.addEventListener("message", (e) => {
                const msg = e.data;
                if (msg.source === "hub" && msg.type === type) {
                  cb(msg.payload);
                }
              });
            }
          };
          resolve(bridge);
        } catch (validationError) {
          console.error("\u274C Invalid Launch Context:", validationError);
          window.parent.postMessage(
            {
              type: MessageType.ERROR,
              source: "game",
              payload: {
                code: "SYS:INVALID_CONTEXT" /* SYS_INVALID_CONTEXT */,
                message: "Launch Context validation failed",
                details: validationError
              }
            },
            targetOrigin
          );
          reject(
            new SdkError(
              "SYS:INVALID_CONTEXT" /* SYS_INVALID_CONTEXT */,
              "Invalid Context",
              validationError
            )
          );
        }
      }
    };
    window.addEventListener("message", messageHandler);
    console.log("[GameSDK] Waiting for Hub...");
    handshakeInterval = setInterval(() => {
      sendMessage(MessageType.HELLO);
    }, 200);
    setTimeout(() => {
      if (!isConnected) {
        clearInterval(handshakeInterval);
        window.removeEventListener("message", messageHandler);
        console.warn(
          "\u26A0\uFE0F [GameSDK] Connection timed out. Switching to DEV/MOCK mode."
        );
        const mockBridge = {
          context: MOCK_CONTEXT,
          send: (type, payload) => {
            console.log(
              `%c[MockBridge \u{1F4E4}] ${type}`,
              "color: green; font-weight: bold",
              payload
            );
          },
          on: (type, cb) => {
            console.log(
              `%c[MockBridge \u{1F442}] Registered listener: ${type}`,
              "color: blue"
            );
            listeners.set(type, cb);
          }
        };
        resolve(mockBridge);
      }
    }, timeoutMs);
  });
}

// src/bridge/hubBridge.ts
var HandshakeHubBridge = class {
  constructor(iframeWindow, context, targetOrigin = "*") {
    __publicField(this, "iframeWindow");
    __publicField(this, "context");
    __publicField(this, "isGameReady", false);
    __publicField(this, "targetOrigin");
    this.iframeWindow = iframeWindow;
    this.context = context;
    this.targetOrigin = targetOrigin;
    this.setupListener();
  }
  setupListener() {
    window.addEventListener("message", (event) => {
      if (this.targetOrigin !== "*" && event.origin !== this.targetOrigin) {
        return;
      }
      const data = event.data;
      if (data?.source !== "game") return;
      switch (data.type) {
        case MessageType.HELLO:
          this.handleGameHello();
          break;
        case MessageType.READY:
          this.handleGameReady();
          break;
      }
    });
  }
  handleGameHello() {
    const envelope = {
      source: "hub",
      type: MessageType.INIT,
      payload: this.context,
      sdkVersion: "1.0.0",
      sessionId: this.context.sessionId,
      gameId: this.context.gameId,
      lessonId: this.context.lessonId,
      gameVersion: this.context.gameVersion
    };
    console.log("[HubSDK] Sending INIT to Game");
    this.iframeWindow.postMessage(envelope, this.targetOrigin);
  }
  handleGameReady() {
    this.isGameReady = true;
    console.log("[HubSDK] Game is fully READY");
  }
  // -------------------- Game Control Methods --------------------
  init(context) {
    this.context = context;
  }
  pause() {
    this.send(MessageType.PAUSE);
  }
  resume() {
    this.send(MessageType.RESUME);
  }
  start() {
    this.send(MessageType.START);
  }
  quit() {
    this.send(MessageType.QUIT);
  }
  resize(width, height, dpr = 1) {
    this.send(MessageType.RESIZE, { width, height, dpr });
  }
  setState(state) {
    this.send(MessageType.SET_STATE, state);
  }
  dispose() {
  }
  getGameReadyStatus() {
    return this.isGameReady;
  }
  send(type, payload) {
    const envelope = {
      source: "hub",
      type,
      payload,
      sdkVersion: "1.0.0",
      sessionId: this.context.sessionId,
      gameId: this.context.gameId,
      lessonId: this.context.lessonId,
      gameVersion: this.context.gameVersion
    };
    this.iframeWindow.postMessage(envelope, this.targetOrigin);
  }
};
function createHubBridge(config, _onEvent) {
  const dummyContext = {
    sessionId: "dummy",
    gameId: "dummy",
    mode: "dev",
    locale: "vi",
    difficulty: "medium",
    player: { name: "Student" },
    soundEnabled: true
  };
  return new HandshakeHubBridge(
    config.iframe.contentWindow,
    dummyContext,
    config.targetOrigin
  );
}

// src/manifest.ts
function isObj(v) {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}
function validateManifest(json) {
  const errors = [];
  const warnings = [];
  function err(path, code, message) {
    errors.push({ path, code, message });
  }
  function warn(path, code, message) {
    warnings.push({ path, code, message });
  }
  if (!isObj(json)) {
    return {
      ok: false,
      errors: [
        {
          path: "",
          code: "type_mismatch",
          message: "Manifest must be an object"
        }
      ],
      warnings: []
    };
  }
  if (typeof json.id !== "string" || !json.id)
    err("id", "required", "Missing or invalid id");
  if (typeof json.title !== "string" || !json.title)
    err("title", "required", "Missing or invalid title");
  if (typeof json.version !== "string")
    err("version", "required", "Missing version");
  else if (!/^\d+\.\d+\.\d+/.test(json.version))
    warn("version", "format", "Version should be semver (x.y.z)");
  if (json.contractVersion !== void 0) {
    if (json.contractVersion !== CONTRACT_VERSION) {
      if (typeof json.contractVersion !== "string")
        err("contractVersion", "type", "Must be string");
    }
  }
  const validTypes = [
    "quiz",
    "drag_drop",
    "trace",
    "classify",
    "memory",
    "custom"
  ];
  if (!json.type || typeof json.type !== "string") {
    err("type", "required", "Missing game type");
  } else if (!validTypes.includes(json.type)) {
    err("type", "invalid", `Type must be one of: ${validTypes.join(", ")}`);
  }
  if (json.subject !== void 0) {
    const validSubjects = [
      "math",
      "vietnamese",
      "english",
      "logic",
      "science",
      "art",
      "music",
      "pe"
    ];
    if (typeof json.subject !== "string" || !validSubjects.includes(json.subject)) {
      err(
        "subject",
        "invalid",
        `Subject must be one of: ${validSubjects.join(", ")}`
      );
    }
  }
  if (json.grade !== void 0) {
    if (!Array.isArray(json.grade)) {
      err("grade", "type", "Grade must be an array");
    } else {
      const validGrades = [
        "pre-k",
        "k",
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
        "7",
        "8",
        "9",
        "10",
        "11",
        "12"
      ];
      json.grade.forEach((g, i) => {
        if (typeof g !== "string" || !validGrades.includes(g)) {
          err(`grade[${i}]`, "invalid", `Invalid grade: ${g}`);
        }
      });
    }
  }
  if (json.tags !== void 0) {
    if (!Array.isArray(json.tags)) {
      err("tags", "type", "Tags must be an array of strings");
    } else {
      json.tags.forEach((t, i) => {
        if (typeof t !== "string")
          err(`tags[${i}]`, "type", "Tag must be string");
      });
    }
  }
  if (typeof json.entryUrl !== "string" || !json.entryUrl) {
    err("entryUrl", "required", "Missing entryUrl");
  }
  if (json.thumbnailUrl !== void 0) {
    if (typeof json.thumbnailUrl !== "string")
      err("thumbnailUrl", "type", "thumbnailUrl must be string");
  } else {
    warn("thumbnailUrl", "missing", "Recommended to add thumbnailUrl");
  }
  if (json.difficulty === void 0) {
    err("difficulty", "required", "Missing difficulty capability list");
  } else {
    if (!Array.isArray(json.difficulty)) {
      err("difficulty", "type", "difficulty must be an array");
    } else {
      const validDiffs = ["easy", "medium", "hard"];
      if (json.difficulty.length === 0)
        warn("difficulty", "empty", "No difficulty levels specified");
      json.difficulty.forEach((d, i) => {
        if (typeof d !== "string" || !validDiffs.includes(d)) {
          err(`difficulty[${i}]`, "invalid", `Invalid difficulty: ${d}`);
        }
      });
    }
  }
  return {
    ok: errors.length === 0,
    errors,
    warnings,
    value: errors.length === 0 ? json : void 0
  };
}

// src/converters/quiz.ts
function convertQuiz(raw) {
  const warnings = [];
  const result = {};
  const correct = typeof raw.correctCount === "number" ? raw.correctCount : void 0;
  const total = typeof raw.totalQuestions === "number" ? raw.totalQuestions : void 0;
  if (correct !== void 0 && total !== void 0 && total > 0) {
    result.accuracy = correct / total;
    result.completion = 1;
    const answered = typeof raw.answeredCount === "number" ? raw.answeredCount : total;
    result.completion = answered / total;
  } else {
    if (typeof raw.score === "number" && typeof raw.maxScore === "number" && raw.maxScore > 0) {
      result.accuracy = raw.score / raw.maxScore;
    }
  }
  return { fields: result, warnings };
}

// src/normalization.ts
function normalizeResult(gameType, raw, context, _manifest) {
  const warnings = [];
  const rawObj = raw && typeof raw === "object" ? raw : {};
  let durationMs = 0;
  if (typeof rawObj.durationMs === "number") durationMs = rawObj.durationMs;
  else if (typeof rawObj.timeTaken === "number")
    durationMs = rawObj.timeTaken;
  else warnings.push("Missing durationMs");
  let score = 0;
  if (typeof rawObj.score === "number") score = rawObj.score;
  else warnings.push("Missing score");
  let attemptId = "";
  if (typeof rawObj.attemptId === "string" && rawObj.attemptId) {
    attemptId = rawObj.attemptId;
  } else if (context?.sessionId) {
    attemptId = context.sessionId;
  } else {
    attemptId = "manual_" + Date.now();
    warnings.push("Missing attemptId, generated fallback");
  }
  let partial = {};
  if (gameType === "quiz") {
    const res = convertQuiz(rawObj);
    partial = res.fields;
    warnings.push(...res.warnings);
  }
  const completion = isFiniteNumber(partial.completion) ? partial.completion : isFiniteNumber(rawObj.completion) ? rawObj.completion : 1;
  const accuracy = isFiniteNumber(partial.accuracy) ? partial.accuracy : isFiniteNumber(rawObj.accuracy) ? rawObj.accuracy : 0;
  const finalCompletion = Math.min(1, Math.max(0, completion));
  const finalAccuracy = Math.min(1, Math.max(0, accuracy));
  const resultPayload = rawObj.resultPayload && typeof rawObj.resultPayload === "object" ? rawObj.resultPayload : rawObj;
  const body = {
    attemptId,
    durationMs,
    score,
    completion: finalCompletion,
    accuracy: finalAccuracy,
    resultPayload
  };
  return { body, warnings };
}
function isFiniteNumber(v) {
  return typeof v === "number" && Number.isFinite(v);
}

exports.CONTRACT_VERSION = CONTRACT_VERSION;
exports.ErrorCode = ErrorCode;
exports.GameResultSchema = GameResultSchema;
exports.HandshakeHubBridge = HandshakeHubBridge;
exports.HubBridge = HandshakeHubBridge;
exports.LaunchContextSchema = LaunchContextSchema;
exports.MAX_MESSAGE_BYTES = MAX_MESSAGE_BYTES;
exports.MessageType = MessageType;
exports.SdkError = SdkError;
exports.clamp01 = clamp01;
exports.connectToHub = connectToHub;
exports.createHubBridge = createHubBridge;
exports.estimateJsonBytes = estimateJsonBytes;
exports.isGameEventType = isGameEventType;
exports.isHubCommandType = isHubCommandType;
exports.isRecord = isRecord;
exports.isValidEnvelope = isValidEnvelope;
exports.makeEnvelope = makeEnvelope;
exports.normalizeResult = normalizeResult;
exports.validateManifest = validateManifest;
exports.validateMessageSize = validateMessageSize;
